---
title: "ml21_fig_prep_v19"
author: "Shyam Saladi"
date: "09/16/2016"
output: html_document
---
```{r}
library(datamart)

library(ggplot2)
library(ggbeeswarm)
library(ggthemes)
library(cowplot)
library(corrplot)
library(viridis)
theme_set(theme_cowplot(font_size=7))

library(gtable)
library(grid)
library(gridExtra)
library(gridGraphics)
library(scales)
library(RColorBrewer)

library(reshape2)
library(plyr)
library(dplyr)
library(tidyr)

library(boot)
library(caret)

library(pROC)
library(locfit)

library(foreach)
library(doMC)
registerDoMC(cores=10)
```

# figure preparation and helper functions
```{r}
specify_decimal <- function(x, k) format(round(x, k), nsmall=k)

substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}

# rounds a decimal number to 3 places (by default) and then formats it as a percentage 
rnd_pct <- function(x, digits = 3, ...) {
    require(scales)
    # cast into character then numeric in case x is a factor
    percent(round(as.numeric(as.character(x)), digits = digits, ...))
}

perc.rank <- function(x) trunc(rank(x))/length(x)
rem_extrema <- function(x) {
    x <- data.frame(V1 = x, V2 = x)
    x[x$V1 != min(x$V1) & x$V1 != max(x$V1),]$V1
}

my_roc <- function(...) {
    require(pROC)
    require(dplyr)
    pROC_outcome <- roc(...)
    pROC_outcome$ppv <- foreach(thisthreshold = pROC_outcome$thresholds, .combine='c') %dopar% {
        truepos <- sum(pROC_outcome$cases >= thisthreshold)
        predpos <- sum(c(pROC_outcome$cases, pROC_outcome$controls) >= thisthreshold)
        truepos/predpos
    }
    pROC_outcome$npv <- foreach(thisthreshold = pROC_outcome$thresholds, .combine='c') %dopar% {
        trueneg <- sum(pROC_outcome$controls <= thisthreshold)
        predneg <- sum(c(pROC_outcome$cases, pROC_outcome$controls) <= thisthreshold)
        trueneg/predneg
    }
    pROC_outcome$threshold_percentiles <- percent_rank(pROC_outcome$thresholds)
    pROC_outcome$min_ppv <- length(pROC_outcome$cases) / (length(pROC_outcome$cases) + length(pROC_outcome$controls))
    pROC_outcome$min_npv <- length(pROC_outcome$controls) / (length(pROC_outcome$cases) + length(pROC_outcome$controls))
#    pROC_outcome$pos_count <- length(pROC_outcome$cases)
    pROC_outcome[c("thresholds", "threshold_percentiles", "sensitivities","specificities","ppv", "min_ppv", "npv", "min_npv")]
}

getname <- function(vector, token = 1) {
    vector <- as.character(vector)
    unlist(lapply(vector, function(x) strsplit(x, "[[:space:]]|(?=[|])", perl=TRUE)[[1]][token]))
}

fd_bw <- function (x) { 2 * IQR(x, na.rm = TRUE) * length(x)^(-1/3) }

level_ranks <- function(x, extra_starting_level = FALSE) {
    x <- as.numeric(factor(x))
    x_levels <- unique(x)
    
    if(extra_starting_level) {
        x_levels <- c(min(x_levels)-1, x_levels)
    }
    map_df <- data.frame(old = x_levels, 
                         new = percent_rank(x_levels))
    map_df$new[match(x, map_df$old)]
}

# modified from from asbio::ConDis.matrix  http://www.inside-r.org/packages/cran/asbio/docs/ConDis.matrix
ConDis.matrix2 <- function (Y1, Y2) 
{
    n <- length(Y1)
    m <- as.data.frame(matrix(nrow = n, ncol = n, dimnames = list(seq(1, 
                                                                      n), seq(1, n))))
    foreach (i=1:n, .combine = 'c') %do% {
        foreach (j=i:n, .combine = 'c') %do% { # changed bounds from 1:n to i:n
            ifelse((Y1[i] > Y1[j] & Y2[i] > Y2[j]) | (Y1[i] < Y1[j] & Y2[i] < Y2[j]), 1,
                   ifelse((Y1[i] < Y1[j] & Y2[i] > Y2[j]) | (Y1[i] > Y1[j] & Y2[i] < Y2[j]), -1, 0))
        }
    }
#    m[upper.tri(m, diag = TRUE)] <- NA
#    m
}

# to make alpha only pertain to fill (not outline)
# from http://stackoverflow.com/questions/34754357/fill-transparency-with-geom-violin/34762360#34762360
GeomPolygon2 <- ggproto("GeomPolygon2", Geom,
                        draw_panel = function(data, panel_scales, coord) {
                          n <- nrow(data)
                          if (n == 1) return(zeroGrob())
                          munched <- coord_munch(coord, data, panel_scales)
                          munched <- munched[order(munched$group), ]
                          first_idx <- !duplicated(munched$group)
                          first_rows <- munched[first_idx, ]
                          ggplot2:::ggname("geom_polygon",
                                           polygonGrob(munched$x, munched$y, default.units = "native",
                                                       id = munched$group,
                                                       gp = gpar(
                                                         col = first_rows$colour,
                                                         fill = alpha(first_rows$fill, first_rows$alpha),
                                                         lwd = first_rows$size * .pt,
                                                         lty = first_rows$linetype
                                                       )
                                           )
                          )
                        },
                        default_aes = aes(colour = "NA", fill = "grey20", size = 0.5, linetype = 1,
                                          alpha = NA),
                        handle_na = function(data, params) {
                          data
                        },
                        required_aes = c("x", "y"),
                        draw_key = draw_key_polygon
)

`%||%` <- function (a, b) 
{
  if (!is.null(a)) 
    a
  else b
}

GeomViolin2 <- ggproto("GeomViolin", Geom,
                       setup_data = function(data, params) {
                         data$width <- data$width %||%
                           params$width %||% (resolution(data$x, FALSE) * 0.9)
                         plyr::ddply(data, "group", transform,
                                     xmin = x - width / 2,
                                     xmax = x + width / 2
                         )
                       },

                       draw_group = function(self, data, ..., draw_quantiles = NULL) {
                         data <- transform(data,
                                           xminv = x - violinwidth * (x - xmin),
                                           xmaxv = x + violinwidth * (xmax - x)
                         )
                         newdata <- rbind(
                           plyr::arrange(transform(data, x = xminv), y),
                           plyr::arrange(transform(data, x = xmaxv), -y)
                         )
                         newdata <- rbind(newdata, newdata[1,])
                         if (length(draw_quantiles) > 0) {
                           stopifnot(all(draw_quantiles >= 0), all(draw_quantiles <= 1))
                           quantiles <- create_quantile_segment_frame(data, draw_quantiles)
                           aesthetics <- data[
                             rep(1, nrow(quantiles)),
                             setdiff(names(data), c("x", "y")),
                             drop = FALSE
                             ]
                           both <- cbind(quantiles, aesthetics)
                           quantile_grob <- GeomPath$draw_panel(both, ...)
                           ggplot2:::ggname("geom_violin", grobTree(
                             GeomPolygon2$draw_panel(newdata, ...),
                             quantile_grob)
                           )
                         } else {
                           ggplot2:::ggname("geom_violin", GeomPolygon2$draw_panel(newdata, ...))
                         }
                       },
                       draw_key = draw_key_polygon,
                       default_aes = aes(weight = 1, colour = "grey20", fill = "white", size = 0.5,
                                         alpha = NA, linetype = "solid"),
                       required_aes = c("x", "y")
)

geom_violin2 <- function(mapping = NULL, data = NULL, stat = "ydensity",
                         draw_quantiles = NULL, position = "dodge",
                         trim = TRUE, scale = "area",
                         na.rm = FALSE, show.legend = NA, inherit.aes = TRUE,
                         ...) {
  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomViolin2,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      trim = trim,
      scale = scale,
      draw_quantiles = draw_quantiles,
      na.rm = na.rm,
      ...
    )
  )
}

# http://stackoverflow.com/a/34859307/2320823
GeomStepHist <- ggproto("GeomStepHist", GeomPath,
                        required_aes = c("x"),
                        
                        draw_panel = function(data, panel_scales, coord, direction) {
                            data <- as.data.frame(data)[order(data$x), ]
                            
                            n <- nrow(data)
                            i <- rep(1:n, each=2)
                            newdata <- rbind(
                                transform(data[1, ], x=x - width/2, y=0),
                                transform(data[i, ], x=c(rbind(data$x-data$width/2, data$x+data$width/2))),
                                transform(data[n, ], x=x + width/2, y=0)
                            )
                            rownames(newdata) <- NULL
                            
                            GeomPath$draw_panel(newdata, panel_scales, coord)
                        }
)

geom_step_hist <- function(mapping = NULL, data = NULL, stat = "bin",
                           direction = "hv", position = "stack", na.rm = FALSE, 
                           show.legend = NA, inherit.aes = TRUE, ...) {
    layer(
        data = data,
        mapping = mapping,
        stat = stat,
        geom = GeomStepHist,
        position = position,
        show.legend = show.legend,
        inherit.aes = inherit.aes,
        params = list(
            direction = direction,
            na.rm = na.rm,
            ...
        )
    )
}

change_cell <- function(table, row, col, name="core", tnew=NA, bnew=NA, lnew=NA, rnew=NA, znew=TRUE) {
    l <- table$layout
    
    ind_fg = which(l$t==row & l$l==col & l$name==paste0(name, '-fg'))
    ind_bg = which(l$t==row & l$l==col & l$name==paste0(name, '-bg'))

    if(!is.na(tnew)) {
        table$layout[ind_fg, 't'] <- tnew
        table$layout[ind_bg, 't'] <- tnew
    }
    if(!is.na(bnew)) {
        table$layout[ind_fg, 'b'] <- bnew
        table$layout[ind_bg, 'b'] <- bnew
    }
    if(!is.na(lnew)) {
        table$layout[ind_fg, 'l'] <- lnew
        table$layout[ind_bg, 'l'] <- lnew
    }
    if(!is.na(rnew)) {
        table$layout[ind_fg, 'r'] <- rnew
        table$layout[ind_bg, 'r'] <- rnew
    }
    if(znew) {
        zpos <- max(p2_auc_table$layout$z)
        table$layout[ind_fg, 'z'] <- zpos+1
        table$layout[ind_bg, 'z'] <- zpos
    }
    table
}
```


# p1 - plot the daley normalized activity vs svmrank score
# fluman norm activity vs svmrank score
# ROCs with cutoffs
```{r}
ecoli_outcomes_exp <- read.csv("training/ecoli_outcomes_exp.csv", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

daley_ml21 <- ecoli_outcomes_exp[! ecoli_outcomes_exp$grouping %in% c('folded.exp','folded.exp.1','folded.exp.2'),]
daley_ml21$scores <- read.table("training/daley_only_raw_exp_ml21.full.4.train.predictions", header = FALSE)$V1

daley_outcomes <- read.csv("training/AllPlates_161104_with_raw.csv", header = TRUE)
daley_outcomes <- filter(daley_outcomes, Keep. == 1) %>% 
    gather(measurement, value, Val1:OD600_Norm4) %>%
    mutate(grouping = factor(paste(groupid, measurement, sep="|")),
           outcome = as.numeric(value)
           ) %>%
    filter(!is.na(outcome))

daley_outcomes <- daley_outcomes[tolower(daley_outcomes$ID) %in% daley_ml21$ID,]
daley_outcomes <- daley_outcomes[daley_outcomes$measurement %in% c('ValNorm1','ValNorm2','ValNorm3', 'ValNorm4'),]
daley_outcomes$ID <- tolower(as.character(daley_outcomes$ID))
daley_outcomes$scores <- join(daley_outcomes, daley_ml21, by = "ID", type = "left", match = "first")$scores

daley_outcomes <- rename(daley_outcomes,  
                         cterm = Cterm_new,
                         ml21 = scores)

## Gather outcomes from Fluman data
fluman_ml21 <- ecoli_outcomes_exp[ecoli_outcomes_exp$grouping %in% c('folded.exp','folded.exp.1','folded.exp.2'),]
fluman_ml21$scores <- read.table("training/daley_only_raw_exp_ml21.full.4.fluman.predictions", header = FALSE)$V1
fluman_ml21 <- select(fluman_ml21, -value)

## only keep those genes that were trained on
fluman_outcomes <- read.csv("training/fluman_data_all.csv", header = TRUE)
fluman_outcomes <- gather(fluman_outcomes, measurement, outcome, whole.cell.fluorescence:OD600.at.harvest) %>%
    filter(!is.na(outcome))
fluman_outcomes <- fluman_outcomes[tolower(fluman_outcomes$name) %in% fluman_ml21$ID,]
fluman_outcomes <- fluman_outcomes[fluman_outcomes$measurement %in% c('folded.exp','folded.exp.1','folded.exp.2'),]
fluman_outcomes$ID <- tolower(as.character(fluman_outcomes$name))

fluman_outcomes <- join(fluman_outcomes, fluman_ml21, by = "ID", type = "left", match = "first")

fluman_outcomes <- mutate(fluman_outcomes, cterm = "fluman", grouping = measurement) %>% 
    rename(ml21 = scores)

## Gather both datasets into the same dataframe with matching columns
ecoli_training <- rbind(select(daley_outcomes, ID, cterm, measurement, outcome, ml21, grouping),
                        select(fluman_outcomes, ID, cterm, measurement, outcome, ml21, grouping))
rm(daley_outcomes, daley_ml21, fluman_outcomes, fluman_ml21)

ecoli_training$grouping <- factor(as.character(ecoli_training$grouping))

# need to calculate AUCs for each activity value but within the "in" and "out" groupings
ecoli_auc <- foreach(thiscterm = unique(ecoli_training$cterm), .combine='rbind') %dopar% {
    thisgroup <- filter(ecoli_training, cterm == thiscterm)

    # within each grouping and go through each activity
    foreach(thisthreshold = unique(rem_extrema(thisgroup$outcome)), .combine='rbind') %dopar% {
        # take all activities greater than the current one as true
        thisresponse <- thisgroup$outcome >= thisthreshold
        data.frame(
            cterm = thiscterm,
            response_threshold = thisthreshold,
            proportion = sum(thisresponse) / length(thisresponse), # proportion of positives/negatives at this activity
            ci = ci.auc(response = thisresponse, predictor = thisgroup$ml21, direction="<", method = "delong"),
            ci_labels = c("lower95", "auc", "upper95")
        )
    }
}
ecoli_auc <- dcast(ecoli_auc, cterm + response_threshold + proportion ~ ci_labels, value.var = "ci")
ecoli_auc$response_threshold_percentile <- perc.rank(ecoli_auc$response_threshold)

ecoli_roc <- foreach(thiscterm = unique(ecoli_training$cterm), .combine='rbind') %dopar% {
    thisgroup <- filter(ecoli_training, cterm == thiscterm)
    # within each grouping and go through each activity
    foreach (thispercentile = c(0:99), .combine='rbind') %dopar% {
        # take all activities greater than the current one as true
        thisquantile <- quantile(thisgroup$outcome, probs = thispercentile/100, na.rm = TRUE, names = FALSE)
        thisresponse <- thisgroup$outcome > thisquantile
        
        thisci <- ci.auc(response = thisresponse, predictor = thisgroup$ml21, direction = "<")
        
        data.frame(
            cterm = thiscterm,
            quantile = thisquantile,
            percentile = thispercentile,
            lower95 = thisci[[1]], 
            auc = thisci[[2]],
            upper95 = thisci[[3]],
            count = nrow(thisgroup),
            pos_count = sum(thisresponse),
            my_roc(response = thisresponse, predictor = thisgroup$ml21, direction = "<")
            )
    }
}

ecoli_training$cterm_rank <- NA
ecoli_training$cterm_rank_bin <- NA
ecoli_training$cterm_rank_percent_bin <- NA
ecoli_training$outcome_rank <- NA

ecoli_auc$response_threshold_percentile <- NA
for(thiscterm in unique(ecoli_training$cterm)) {
    
    thisrank <- rank(ecoli_training[ecoli_training$cterm == thiscterm,]$ml21)
    
    ecoli_training[ecoli_training$cterm == thiscterm,]$cterm_rank <- thisrank
    ecoli_training[ecoli_training$cterm == thiscterm,]$cterm_rank_bin <- floor(thisrank/10) *10
    ecoli_training[ecoli_training$cterm == thiscterm,]$cterm_rank_percent_bin <- percent_rank(floor(thisrank/10) *10)
    
    ecoli_training[ecoli_training$cterm == thiscterm,]$outcome_rank <-
        percent_rank(ecoli_training[ecoli_training$cterm == thiscterm,]$outcome)
    
    ecoli_auc[ecoli_auc$cterm == thiscterm,]$response_threshold_percentile <-
        percent_rank(ecoli_auc[ecoli_auc$cterm == thiscterm,]$response_threshold)
}
rm(thiscterm)

ecoli_training$cterm_rank_bin <- floor(ecoli_training$cterm_rank/10)*10

ecoli_auc$cterm <- factor(ecoli_auc$cterm, 
                          labels = c("in", "out", "fluman"), 
                          levels = c("in", "out", "fluman"))


ecoli_roc$cterm <- factor(ecoli_roc$cterm, 
                          labels = c("C-terminus Inside (GFP activity)", "C-terminus Outside (PhoA activity)", 
                                     "C-terminus Inside (Folded protein"), 
                          levels = c("in", "out", "fluman"))

ecoli_cor <- foreach (thisgrouping = unique(ecoli_training$grouping), .combine='rbind') %dopar% {
    thisgroup <- filter(ecoli_training, grouping == thisgrouping) #"folded.exp") 
    pairs <- ConDis.matrix2(thisgroup$ml21, thisgroup$outcome)
    data.frame(group = thisgrouping,
               cterm = unique(thisgroup$cterm),
               table(pairs))
}
ecoli_cor <- dcast(ecoli_cor, cterm + group ~ pairs, value.var = "Freq")
colnames(ecoli_cor) <- c("cterm", "group", "disconcor", "invalid", "concor")
ecoli_cor <- mutate(ecoli_cor, total = disconcor + concor) %>%
    group_by(cterm) %>%
    summarize(concor = sum(concor, na.rm = TRUE),
              disconcor = sum(disconcor, na.rm = TRUE),
              total = sum(total, na.rm = TRUE),
              kendall = (concor - disconcor)/total)

# make table from here:
ecoli_stats <- ecoli_training %>% 
    group_by(cterm) %>%
    summarize(plates = length(unique(grouping)),
              genes = length(unique(ID)),
              observations = length(ID)) %>%
    inner_join(ecoli_cor, by = c("cterm" = "cterm")) %>%
    select(-disconcor, -concor) %>%
    rename(total_pairs = total) %>% 
    data.frame

rownames(ecoli_stats) <- ecoli_stats$cterm
ecoli_stats <- select(ecoli_stats, -cterm)

ecoli_stats$kendall <- round(ecoli_stats$kendall, 3)
colnames(ecoli_stats) <- c("Plates", "Genes", "Observations", # "Concordant Pairs", "Disconcordant Pairs", 
                           "Total~Pairs", "Kendall's"~tau)

table_theme <-  ttheme_default(
    core = list(fg_params=list(fontsize = 6)),
    colhead = list(fg_params=list(fontsize = 6, parse=TRUE)),
    rowhead = list(fg_params=list(fontsize = 6, parse=TRUE)),
    padding = unit(c(2, 3), "mm"))

tab <- tableGrob(data.frame(t(ecoli_stats)), theme = table_theme, 
                 cols = c("C-terminal Cytoplasmic\n(GFP)", 
                          "C-terminal Periplasmic\n(PhoA)", 
                          "C-terminal Cytoplasmic\n(GFP, folded protein)"))
header <- tableGrob(data.frame(t(ecoli_stats))[1,1:2],
                    theme = table_theme,
                    cols=c("Daley, Rapp, et al., 2005", "Fluman, et al., 2014")) 

p1_table <- gridExtra::combine(header[1,], tab, along=2)
p1_table$layout[1:6 , c("l", "r")] <- list(c(2, 4), c(3, 4))

# chg_attr <- function(table, row, col, name="core-fg", append = TRUE, ...) {
#     l <- table$layout
#     ind <- which(l$t==row & l$l==col & l$name==name)
#     
#     table$grobs[ind][[1]][["gp"]] <- c(table$grobs[ind][[1]][["gp"]], ...)
#     table
# }

ind <- find_cell(p1_table, 2, 2, "colhead-fg")
p1_table$grobs[ind][[1]][["gp"]]$col <- brewer.pal(3, "Dark2")[[1]]
ind <- find_cell(p1_table, 2, 3, "colhead-fg")
p1_table$grobs[ind][[1]][["gp"]]$col <- brewer.pal(3, "Dark2")[[2]]
ind <- find_cell(p1_table, 2, 4, "colhead-fg")
p1_table$grobs[ind][[1]][["gp"]]$col <- brewer.pal(3, "Dark2")[[3]]


# correlation metrics
ecoli_xy_cor <- ecoli_training %>% group_by(cterm) %>%
    summarize(count = length(outcome),
              spearman = cor(outcome, ml21, method = "spearman"),
              # lower95 = specify_decimal(spearman_95ci(spearman, count), k=2)[1],
              # upper95 = specify_decimal(spearman_95ci(spearman, count), k=2)[2],
              spearman = specify_decimal(spearman, k=2),
              # ci = paste0(spearman, " (", lower95, "-", upper95, ")"),
              x = min(ml21)+(max(ml21)-min(ml21))*(6/20), 
              ylab = max(outcome)*(1-1/50),
              y = ylab-max(outcome)*(1/20)
    )

ecoli_daley_fluman <- filter(ecoli_training, cterm != "out") %>% group_by(cterm, ID) %>% 
    summarize(outcome_max = max(outcome, na.rm = TRUE),
              outcome_avg = mean(outcome, na.rm = TRUE),
              outcome_min = min(outcome, na.rm = TRUE))
ecoli_daley_fluman <- cbind(
    dcast(ecoli_daley_fluman, ID ~ cterm, value.var = "outcome_min"),
    dcast(ecoli_daley_fluman, ID ~ cterm, value.var = "outcome_avg"),
    dcast(ecoli_daley_fluman, ID ~ cterm, value.var = "outcome_max"))
colnames(ecoli_daley_fluman) <- c("id", "daley_min", "fluman_min", "id2", "daley_avg", "fluman_avg", "id3", "daley_max", "fluman_max")
ecoli_daley_fluman <- filter(ecoli_daley_fluman, !is.na(fluman_min), !is.na(daley_min))

ecoli_daley_fluman_cor <- cor(ecoli_daley_fluman$fluman_avg, ecoli_daley_fluman$daley_avg, method="spearman")
#ecoli_daley_fluman_cor_ci <- specify_decimal(spearman_95ci(ecoli_daley_fluman_cor, nrow(ecoli_daley_fluman)), k=2)
```

# bootstrapping and p-value calculation code in training_bootstraps_with_bca_ci.R
# Just load RData file
```{r}
# save.image("training_bootstraps_with_bca_ci_envinput.RData")
load("training_bootstraps_with_bca_ci_summary.RData")
```

# p1 - ecoli training plots
```{r}
# all possible cutoffs with AUCs
p1_ecoli_auc <- ggplot(ecoli_auc, aes(x=response_threshold_percentile*100, color = cterm)) + 
    geom_hline(aes(yintercept=50), linetype = "dashed", color = "darkgrey") + 
    geom_line(aes(y = auc*100)) + 
#    geom_line(aes(y = lower95)) + geom_line(aes(y = upper95)) +
    scale_y_continuous(expand = c(0, 0), breaks=seq(0,100,10), limits = c(48, 90)) + 
    scale_x_continuous(expand = c(0, 0), breaks=seq(0,100,25), limits = c(0,100)) + 
    ylab("AUC") + 
    xlab("Activity Percentile") + 
    scale_color_manual(values = brewer.pal(3, "Dark2")) +
    theme(legend.position = "None")

ecoli_selected_auc <- filter(ecoli_roc, percentile %in% c(50, 97)) %>%
    mutate(cterm = factor(cterm)) %>% 
    group_by(cterm, percentile) %>% 
    select(cterm, percentile, lower95, auc, upper95, pos_count, count) %>%
    distinct() %>%
    mutate(auc = specify_decimal(auc*100, k = 1),
           lower95 = specify_decimal(lower95*100, k = 1),
           upper95 = specify_decimal(upper95*100, k = 1),
           ci = paste0("'", auc, "'~('", lower95, "'-'", upper95, "')"))

ecoli_selected_counts <- data.frame(percentile = c(50,97), counts = c(-1, -1))
ecoli_selected_counts[ecoli_selected_counts$percentile == 50,]$counts <- 
    paste(filter(ecoli_selected_auc, percentile == 50)$pos_count %>% toString)
ecoli_selected_counts[ecoli_selected_counts$percentile == 97,]$counts <- 
    paste(filter(ecoli_selected_auc, percentile == 97)$pos_count %>% toString)

p1_ecoli_roc <- ggplot(filter(ecoli_roc, percentile %in% c(50, 97))) + 
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "darkgrey") +
    geom_path(aes(x=100-specificities*100, y=sensitivities*100, color = cterm)) +
    geom_text(aes(x = 75, y = 8*(4-as.numeric(cterm))-3, color = cterm, label = ci), 
              size = 5*5/14, parse = TRUE, data = ecoli_selected_auc) +
    geom_text(aes(x = 75, y = 29, label = counts), 
              size = 5*5/14, parse = FALSE, data = ecoli_selected_counts) +
    scale_y_continuous(expand = c(0, 0)) +
    scale_x_continuous(expand = c(0, 0)) +
    xlab("False Positive Rate") + ylab("True Positive Rate") + 
    scale_color_manual(values = brewer.pal(3, "Dark2")) +
    facet_wrap(~percentile, nrow=1, scales = "free")  +
    theme(legend.position = "None",
          aspect.ratio = 1,
          plot.margin = unit(c(0, 0, 0, 0), "mm"),
          strip.background = element_blank())

p1_ecoli_roc <- ggplotGrob(p1_ecoli_roc)
p1_ecoli_roc$grobs$strip_t1$grobs[[1]]$children[[2]]$children[[1]]$label <- expression(50^"th"*~"percentile")
p1_ecoli_roc$grobs$strip_t2$grobs[[1]]$children[[2]]$children[[1]]$label <- expression(97^"th"*~"percentile")


ecoli_xy_cor$lower95_boot <- -1
ecoli_xy_cor$upper95_boot <- -1
ecoli_xy_cor[ecoli_xy_cor$cterm == "in", c('lower95_boot', 'upper95_boot')] <- ecoli_xy_boot_in_ci$bca[c(4,5)]
ecoli_xy_cor[ecoli_xy_cor$cterm == "out", c('lower95_boot', 'upper95_boot')] <- ecoli_xy_boot_out_ci$bca[c(4,5)]
ecoli_xy_cor[ecoli_xy_cor$cterm == "fluman", c('lower95_boot', 'upper95_boot')] <- ecoli_xy_boot_fluman_ci$bca[c(4,5)]
ecoli_xy_cor <- mutate(ecoli_xy_cor, 
                       lower95_boot = specify_decimal(lower95_boot, k=2),
                       upper95_boot = specify_decimal(upper95_boot, k=2),
                       ci_boot = paste0("'", spearman, "'~('", lower95_boot, "'-'", upper95_boot, "')")
                       )
ecoli_xy_cor$name <- "name"
ecoli_xy_cor[ecoli_xy_cor$cterm == "in", 'name'] <- "GFP"
ecoli_xy_cor[ecoli_xy_cor$cterm == "fluman", 'name'] <- "GFP, Folded Protein"

p1_ecoli_xy <- ggplot(ecoli_training, aes(x=ml21, y=outcome, color=cterm)) +
    stat_summary(fun.ymin = min, fun.ymax = max, geom = "linerange", size = 0.3, alpha=0.35, color="black") +
    stat_summary(fun.y = mean, geom="point", size = .25, alpha = 0.7) +
    geom_text(aes(x = -2, y = y, color = cterm, label = ci_boot), size = 5*5/14, parse = TRUE, data = ecoli_xy_cor) +
#     geom_text(aes(x=x, y=y*1.08, color=cterm, label=name), size = 5*5/14, parse=FALSE, data = ecoli_xy_cor) +
    scale_y_continuous(expand = c(.01, 0), breaks=c(0:3, 4, 6, 8)) +
    scale_x_continuous(expand = c(.01, 0.01)) +
    scale_color_manual(values = brewer.pal(3, "Dark2")) + 
    ylab("Mean Normalized Activity") +
    xlab(expression("SVM"^"rank"*~"score")) + 
    theme(legend.position = "None",
              strip.text = element_blank(),
              plot.margin = unit(c(0, 0, 0, 0), "mm")) +
    facet_wrap(~cterm, ncol=1, scales = "free_y")

p1_ecoli_xy <- ggplotGrob(p1_ecoli_xy)
p1_ecoli_xy$grobs$axis_l1$children$axis$grobs[[1]]$children[[1]]$label <- c("0", "", "2", "", "4", "6", "8")

ecoli_daley_fluman_cor_label <- specify_decimal(c(ecoli_daley_fluman_cor, 
                                                  ecoli_daley_fluman_boot_ci$bca[4], 
                                                  ecoli_daley_fluman_boot_ci$bca[5]), 
                                                k=2)
ecoli_daley_fluman_cor_label <- paste0("'", ecoli_daley_fluman_cor_label[1], "'~('",
                                       ecoli_daley_fluman_cor_label[2], "'-'",
                                       ecoli_daley_fluman_cor_label[3], "')")

p1_daley_fluman <- ggplot(ecoli_daley_fluman) + 
    geom_segment(aes(x=fluman_min, xend=fluman_max, y=daley_avg, yend=daley_avg), size=0.3, alpha=0.35) +
    geom_segment(aes(x=fluman_avg, xend=fluman_avg, y=daley_min, yend=daley_max), size=0.3, alpha=0.35) +
    geom_point(aes(x=fluman_avg, y=daley_avg), size=0.25) +
    annotate("text", x = 1, y = 7.45, label = ecoli_daley_fluman_cor_label, size = 5*5/14, parse=TRUE) +
    scale_y_continuous(expand = c(0, 0)) +
    scale_x_continuous(expand = c(.01, 0)) + 
    ylab("Normalized GFP activity") +
    xlab("Normalized folded protein") + 
    theme(axis.title.x = element_text(size=7, color=brewer.pal(3, "Dark2")[3]),
              axis.title.y = element_text(size=7, color=brewer.pal(3, "Dark2")[1]),
              plot.margin = unit(c(0, 0, 0, .5), "mm"))

p1_main <- ggdraw() +
    draw_plot(p1_daley_fluman, x = .0, y = .5, width = .23, height = 0.48) + 
    draw_plot(p1_table, x = .21, y = 0.51, width = 1 - .47, height = 0.48) +
    draw_plot(p1_ecoli_xy, x = .75, y = 0, width = .24, height = .98) + 
    draw_plot(plot_grid(p1_ecoli_roc, p1_ecoli_auc, nrow = 1, rel_widths = c(1, 1)), 
              x = 0, y = 0, width = .75, height = .5) + 
    draw_plot_label(letters[1:5], c(0, 0.26, .75, 0, 0.38), c(1, 1, 1, 0.52, 0.52), size = 8)

save_plot(filename = "plots/fig1_w_panels.pdf", 
          plot = p1_main,
          base_width = uconv(183, "mm", "in", "Length"),
          base_height = 3)
```

# p2 - nycomps
```{r}
source("preparation-convenience.R")

# Load features as well as well as scores
nycomps_allstats <- read.csv("large-scale/nycomps_061116.fna.allstats.csv", header = TRUE, as.is = TRUE) %>%
    separate(title, c("del", "orfid"), sep = "\\$|\\.|_") %>%
    mutate(orfno = as.integer(orfid)) %>%
    select(-del, -orfid) %>%
    cbind(read.table("large-scale/nycomps_061116.fna.allstats.ml21", header = FALSE) %>% rename(ml21 = V1)) %>%
    cbind(read.table("large-scale/nycomps_061116.fna.phobiuscterm", header = FALSE) %>% rename(cterm = V1))

# Load outcomes from NYCOMPS and join ml21 by row
nycomps_outcomes <- read.csv("large-scale/nycomps-outcomes-all-actions-raw-dump.csv", header = TRUE) %>%
    left_join(nycomps_allstats %>% select(orfno, ml21), by = "orfno")

# Find out how many we don't have sequence information for
nycomps_outcomes %>% filter(is.na(ml21)) %>% distinct(orfno) %>% nrow
nycomps_outcomes$orfno %>% unique %>% length
# 817 are missing out of 12697 genes (6.434591%)

# As throughout, can only work with those targets for which we have sequences/scores along with the 
nycomps_outcomes <- filter(nycomps_outcomes, !is.na(ml21))

# AUC for all small scale expression trials
nycomps_all_trials <- nycomps_outcomes %>%
    filter(action %in% c(#"Expression by western -",
                                      "Coomassie Negative",
                                      "Coomassie Positive",
                                      "Passed to Large-Scale"),
           grepl('Expressed-', type)) %>%
    mutate(action = factor(as.character(action),
                           levels = c("Expression by western -",
                                      "Coomassie Negative",
                                      "Coomassie Positive",
                                      "Passed to Large-Scale"),
                           labels = c("Expression by western -",
                                      "Coomassie Negative",
                                      "Coomassie Positive",
                                      "Passed to Large-Scale"),
                           ordered = TRUE))


nycomps_ <- nycomps_outcomes %>%
    filter(action %in% c(#"Expression by western -",
                                      "Coomassie Negative",
                                      "Coomassie Positive",
                                      "Passed to Large-Scale"),
           grepl('Expressed-', type)) %>%
    mutate(action = factor(as.character(action),
                           levels = c("Expression by western -",
                                      "Coomassie Negative",
                                      "Coomassie Positive",
                                      "Passed to Large-Scale"),
                           labels = c("Expression by western -",
                                      "Coomassie Negative",
                                      "Coomassie Positive",
                                      "Passed to Large-Scale"),
                           ordered = TRUE))

# Looking at the level of proteins, so only look at distinct outcomes (not retests)
nycomps_largescale <- nycomps_outcomes %>%
    filter(action %in% c("Passed to Large-Scale", "Coomassie Negative"),
           grepl('Expressed-', type)) %>%
    distinct(orfno, ml21, type, action) %>%
    group_by(orfno, ml21, type)
    group_by(orfno, ml21) %>%
    summarize(ntrials = n(),
              pos_count = sum(action %in% "Passed to Large-Scale"),
              class = ifelse(pos_count == 0, "neg", 
                          ifelse(pos_count == ntrials, "pos", "mix")))  %>%
    ungroup %>%
    mutate(class = factor(class,
                          levels = c("neg", "mix", "pos"),
                          labels = c("neg", "mix", "pos"),
                          ordered = TRUE)) %>% 
    as.data.frame

# Looking at the level of proteins, so only look at distinct outcomes (not retests)
nycomps_genes <- nycomps_vectors %>%
    distinct(orfno, ml21, type, action) %>%
    group_by(orfno, ml21) %>%
    summarise(ntrials = n(),
              pos_count = sum(action %in% "Coomassie Positive"),
              class = ifelse(pos_count == 0, "neg", 
                          ifelse(pos_count == ntrials, "pos", "mix")))  %>%
    ungroup %>%
    mutate(class = factor(class,
                          levels = c("neg", "mix", "pos"),
                          labels = c("neg", "mix", "pos"),
                          ordered = TRUE)) %>% 
    as.data.frame
    

nycomps_auc <- rbind(
    calc_auc_roc(nycomps_vectors %>% mutate(vectors = "all"), # %>% filter(action != "Expression by western -"),
                 grouping_col = "vectors",
                 outcome_col = "action"),
    calc_auc_roc(nycomps_vectors %>% mutate(vectors = type),
                 grouping_col = "vectors",
                 outcome_col = "action"),
    calc_auc_roc(nycomps_genes %>% mutate(genes = "all"),
             grouping_col = "genes",
             outcome_col = "class"),
    calc_auc_roc(nycomps_largescale %>% mutate(largescale = "all"),
                 grouping_col = "largescale",
                 outcome_col = "class")) %>%
    rename(metagroup = grouping_col)

nycomps_roc <- rbind(
    calc_auc_roc(nycomps_vectors %>% mutate(vectors = "all"),
                 grouping_col = "vectors",
                 outcome_col = "action",
                 method = "roc"),
    calc_auc_roc(nycomps_genes %>% mutate(genes = "all"),
                 grouping_col = "genes",
                 outcome_col = "class",
                 method = "roc"),
    calc_auc_roc(nycomps_largescale %>% mutate(largescale = "all"),
                 grouping_col = "largescale",
                 outcome_col = "action",
                 method = "roc")) %>%
    rename(metagroup = grouping_col)

nycomps_xscale <- quantile(nycomps_genes$ml21, probs = c(.0025,.9975), names = FALSE)
nycomps_xscale[[1]] <- nycomps_xscale[[1]]-1
nycomps_xscale[[2]] <- nycomps_xscale[[2]]+1
nycomps_breaks <- seq(-6, 6, 2)
```

# nycomps - noRNAss
```{r}
nycomps_pos_scores_noRNAss <- data.frame(
    ml21 = read.table("large-scale/nycomps_only_pos.fna.allstats_noRNAss.csv.ml21", header = FALSE)$V1,
    cterm = read.table("large-scale/nycomps_only_pos_phobius.phobiuscterm", header = FALSE)$V1,
    gene_outcome = "pos",
    row.names = rownames(read.csv("large-scale/nycomps_only_pos.fna.allstats_noRNAss.csv", header = TRUE, row.names = 1))
    )

nycomps_neg_scores_noRNAss <- data.frame(
    ml21 = read.table("large-scale/nycomps_only_neg.fna.allstats_noRNAss.csv.ml21", header = FALSE)$V1,
    cterm = read.table("large-scale/nycomps_only_neg_phobius.phobiuscterm", header = FALSE)$V1,
    gene_outcome = "neg",
    row.names = rownames(read.csv("large-scale/nycomps_only_neg.fna.allstats_noRNAss.csv", header = TRUE, row.names = 1))
    )

nycomps_mixed_scores_noRNAss <- data.frame(
    ml21 = read.table("large-scale/nycomps_mixed.fna.allstats_noRNAss.csv.ml21", header = FALSE)$V1,
    cterm = read.table("large-scale/nycomps_mixed_phobius.phobiuscterm", header = FALSE)$V1,
    gene_outcome = "mix",
    row.names = rownames(read.csv("large-scale/nycomps_mixed.fna.allstats_noRNAss.csv", header = TRUE, row.names = 1))
    )
nycomps_posmix_noRNAss <- rbind(nycomps_pos_scores_noRNAss, nycomps_neg_scores_noRNAss, nycomps_mixed_scores_noRNAss)
rm(nycomps_pos_scores_noRNAss, nycomps_neg_scores_noRNAss, nycomps_mixed_scores_noRNAss)

rownames(nycomps_posmix_noRNAss) <- getname(rownames(nycomps_posmix_noRNAss), 3)
nycomps_posmix_noRNAss$id <- as.integer(rownames(nycomps_posmix_noRNAss))
nycomps_posmix_noRNAss$gene_outcome <- as.character(nycomps_posmix_noRNAss$gene_outcome)
nycomps_posmix_noRNAss[nycomps_posmix_noRNAss$gene_outcome %in% c("mix", "pos"),]$gene_outcome <- "≥1 pos"
nycomps_posmix_noRNAss$gene_outcome <- factor(nycomps_posmix_noRNAss$gene_outcome, 
                              levels = c("neg", "≥1 pos"), 
                              labels = c("neg", "≥1 pos"))

nycomps_vectors_noRNAss <- read.csv("large-scale/nycomps_outcomes.csv", header = TRUE,  stringsAsFactors = FALSE) %>% 
    rename(id = X,
           sum_outcomes = sum) %>%
    gather(plasmid_name, plasmid_outcome, N:MSGC.9_LDAO) %>%
    filter(Analyze == 1) %>%
    select(orfno, id, sum_outcomes, tot_neg, tot_pos, pos_neg, plasmid_name, plasmid_outcome)

nycomps_vectors_noRNAss <- left_join(nycomps_vectors_noRNAss, nycomps_posmix_noRNAss, by = "id") %>%
    filter(!is.na(ml21), plasmid_outcome != 0)

nycomps_roc_noRNAss <- foreach(thisplasmid = unique(nycomps_vectors_noRNAss$plasmid_name), .combine='rbind') %dopar% {
    thisdata <- filter(nycomps_vectors_noRNAss, plasmid_name == thisplasmid)
    thisdata <- transmute(thisdata, 
                          thisresponse = plasmid_outcome == 1,
                          ml21 = ml21)
    df <- data.frame(
        my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
        group = thisplasmid,
        type = "plasmid"
        )
}

nycomps_roc_noRNAss <- 
    rbind(nycomps_roc_noRNAss,
          data.frame(my_roc(gene_outcome == "≥1 pos" ~ ml21, data = nycomps_posmix_noRNAss, direction = "<"), group = "posmix", type = "all"),
          data.frame(my_roc(plasmid_outcome == 1 ~ ml21, data = nycomps_vectors_noRNAss, direction = "<"), group = "vectors", type = "all")
    )

nycomps_auc_noRNAss <- foreach(thisplasmid = unique(nycomps_vectors_noRNAss$plasmid_name), .combine=rbind) %dopar% {
    thisdata <- filter(nycomps_vectors_noRNAss, plasmid_name == thisplasmid)
    thisdata <- transmute(thisdata, 
                          thisresponse = plasmid_outcome == 1,
                          ml21 = ml21)
    df1 <- data.frame(
        ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
        group = thisplasmid,
        type = "plasmid",
        cterm = "all",
        ci_labels = c("lower95", "auc", "upper95"),
        count = nrow(thisdata),
        pos_count = sum(thisdata$thisresponse)
        )
    
    df2 <- foreach(thiscterm = unique(nycomps_vectors_noRNAss$cterm), .combine=rbind) %dopar% {
        thisdata <- filter(nycomps_vectors_noRNAss, plasmid_name == thisplasmid, cterm == thiscterm)
        thisdata <- transmute(thisdata, 
                              thisresponse = plasmid_outcome == 1,
                              ml21 = ml21)
        data.frame(
            ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
            group = thisplasmid,
            type = "plasmid",
            cterm = thiscterm,
            ci_labels = c("lower95", "auc", "upper95"),
            count = nrow(thisdata),
            pos_count = sum(thisdata$thisresponse)
            )
    }
    
    rbind(df1, df2)
}

ci.auc(large_scale ~ ml21, data = nycomps_vectors)

nycomps_auc_noRNAss <- 
    rbind(nycomps_auc_noRNAss,
          data.frame(ml21 = ci.auc(gene_outcome == "≥1 pos" ~ ml21, 
                                   data = nycomps_posmix_noRNAss, direction = "<"),
                    group = "posmix", 
                    type = "all", 
                    cterm = "all",
                    ci_labels = c("lower95", "auc", "upper95"),
                    count = nrow(nycomps_posmix_noRNAss),
                    pos_count = sum(nycomps_posmix_noRNAss$gene_outcome == "≥1 pos")
                    ),
          data.frame(ml21 = ci.auc(gene_outcome == "≥1 pos" ~ ml21, 
                                   data = filter(nycomps_posmix_noRNAss, cterm == "CYTOPLASMIC."), direction = "<"),
                    group = "posmix", 
                    type = "all", 
                    cterm = "CYTOPLASMIC.",
                    ci_labels = c("lower95", "auc", "upper95"),
                    count = nrow( filter(nycomps_posmix_noRNAss, cterm == "CYTOPLASMIC.") ),
                    pos_count = sum(filter(nycomps_posmix_noRNAss, cterm == "CYTOPLASMIC.")$gene_outcome == "≥1 pos")
                    ),
          data.frame(ml21 = ci.auc(gene_outcome == "≥1 pos" ~ ml21, 
                                   data = filter(nycomps_posmix_noRNAss, cterm == "NON-CYTOPLASMIC."), direction = "<"),
                    group = "posmix", 
                    type = "all", 
                    cterm = "NON-CYTOPLASMIC.",
                    ci_labels = c("lower95", "auc", "upper95"),
                    count = nrow( filter(nycomps_posmix_noRNAss, cterm == "NON-CYTOPLASMIC.") ),
                    pos_count = sum(filter(nycomps_posmix_noRNAss, cterm == "NON-CYTOPLASMIC.")$gene_outcome == "≥1 pos")
                    ),
          data.frame(ml21 = ci.auc(plasmid_outcome == 1 ~ ml21, 
                                   data = nycomps_vectors_noRNAss, direction = "<"),
                    group = "vectors", 
                    type = "all", 
                    cterm = "all",
                    ci_labels = c("lower95", "auc", "upper95"),
                    count = nrow(nycomps_vectors_noRNAss),
                    pos_count = sum(nycomps_vectors_noRNAss$plasmid_outcome == 1)
                    ),
          data.frame(ml21 = ci.auc(plasmid_outcome == 1 ~ ml21, 
                                   data = filter(nycomps_vectors_noRNAss, cterm == "CYTOPLASMIC."), direction = "<"),
                    group = "vectors", 
                    type = "all", 
                    cterm = "CYTOPLASMIC.",
                    ci_labels = c("lower95", "auc", "upper95"),
                    count = nrow( filter(nycomps_vectors_noRNAss, cterm == "CYTOPLASMIC.") ),
                    pos_count = sum(filter(nycomps_vectors_noRNAss, cterm == "CYTOPLASMIC.")$plasmid_outcome == 1)
                    ),
          data.frame(ml21 = ci.auc(plasmid_outcome == 1 ~ ml21, 
                                   data = filter(nycomps_vectors_noRNAss, cterm == "NON-CYTOPLASMIC."), direction = "<"),
                    group = "vectors", 
                    type = "all", 
                    cterm = "NON-CYTOPLASMIC.",
                    ci_labels = c("lower95", "auc", "upper95"),
                    count = nrow( filter(nycomps_vectors_noRNAss, cterm == "NON-CYTOPLASMIC.") ),
                    pos_count = sum(filter(nycomps_vectors_noRNAss, cterm == "NON-CYTOPLASMIC.")$plasmid_outcome == 1)
                    )
    )

nycomps_auc_noRNAss <- dcast(nycomps_auc_noRNAss, group + type + cterm + count + pos_count ~ ci_labels, value.var = "ml21")

nycomps_auc_noRNAss_table <- filter(nycomps_auc_noRNAss, cterm == "all") %>% select(-count, -pos_count, -type, -cterm) %>%
    mutate(auc = specify_decimal(auc*100, k=1),
           lower95 = specify_decimal(lower95*100, k=1),
           upper95 = specify_decimal(upper95*100, k=1),
           label = paste0("'", auc, "'~('", lower95, "'-'", upper95, "')"),
           group = factor(as.character(group),
               levels = c("all", "vectors", "N", "C", "C_LDAO", 
                          "MSGC.24", "MSGC.28", "MSGC.7", "MSGC.9", "MSGC.9_LDAO"),
               labels = c("All Conditions", "(N) His-FLAG-TEV-", "(C) -TEV-His", "(C) -TEV-His (LDAO)", 
                          "(24) His-GST-TEV-", "(28) -TEV-His", "(7) His-TEV-", "(9) His-MBP-TEV-", "(9) His-MBP-TEV- (LDAO)")))

write.csv(nycomps_auc_noRNAss, file = "nycomps_nonupack_auc.csv")

table_theme2 <- ttheme_default(
    core = list(fg_params=list(fontsize = 5, fontfamily = "Helvetica", parse=TRUE)),
    colhead = list(fg_params=list(fontsize = 5, fontfamily = "Helvetica")),
    rowhead = list(fg_params=list(fontsize = 5, fontfamily = "Helvetica")),
    padding = unit(c(1, 2), "mm"))

p4_auc_table_nonupack <- tableGrob(nycomps_cterm, theme = table_theme2, 
                 cols = rep(c("Count", "AUC (95% CI)" ), 3),
                 rows = as.character(temp_names),
                 widths = unit(rep(c(.085, 0.165), 3), "npc"),
                 heights = unit(rep(.0001, length(temp_names)), "npc"))


```


# nycomps plotting
```{r}
p2_histogram <- ggplot() + 
    geom_histogram(aes(x=ml21, fill = class != "neg"),
                   binwidth = fd_bw(nycomps_largescale$ml21),
                   alpha=1, position="identity",
                   data=nycomps_largescale) +
    geom_step_hist(aes(x=ml21),
                   color = "#A10F1C",
                   binwidth = fd_bw(nycomps_largescale$ml21),
                   data = filter(nycomps_genes, class != "neg")) +
    geom_vline(xintercept = quantile(nycomps_largescale$ml21, probs = 0.75, names = FALSE), linetype = "dashed") + 
    scale_x_continuous(expand=c(0,0), breaks=nycomps_breaks, limits=nycomps_xscale) +
    scale_y_continuous(expand=c(0,0)) +
    scale_fill_manual(values = c("#A4A5A5","#A10F1C"), guide = guide_legend(title = "Gene of Interest:")) +
    ylab("Count") +
    xlab(expression("SVM"^"rank"*~"score")) + 
    theme(legend.key = element_blank(),
          legend.key.size = unit(0.25, "cm"),
          legend.title = element_blank(),
              legend.position=c(0.2, 0.8), 
              plot.margin=unit(c(0, 0, 0, 0), "in"))

p2_histogram <- ggplotGrob(p2_histogram)
p2_histogram$grobs[[8]]$grobs[[1]][[1]][[7]]$gp$col <- "#A4A5A5"
p2_histogram$grobs[[8]]$grobs[[1]][[1]][[8]]$gp$col <- "#A10F1C"

format_auc_df <- function(auc_df) {
    auc_df %>%
        mutate(auc_x = specify_decimal(auc*100, k = 1),
               lower95_x = specify_decimal(lower95*100, k = 1),
               upper95_x = specify_decimal(upper95*100, k = 1),
               ci = paste0("'", auc_x, "'~('", lower95_x, "'-'", upper95_x, "')"),
               count = paste0(pos_count, "/", count)) %>%
        select(-auc_x, -lower95_x, -upper95_x)
}

p2_posmix_roc_label <- nycomps_auc %>%
    filter(group == "all") %>%
    format_auc_df()

p2_posmix_roc <- 
    
    ggplot() +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "darkgrey") +
    geom_path(aes(x = 1-specificities, y = sensitivities),
              color = "blue",
              data = filter(nycomps_roc,
                            metagroup == "vectors",
                            group == "all")) +
    geom_path(aes(x = 1-specificities, y = sensitivities),
              color = "#A10F1C",
              data = filter(nycomps_roc,
                            metagroup == "largescale",
                            group == "all")) +
    geom_path(aes(x = 1-specificities, y = sensitivities),
              color = "#A10F1C",
              linetype = "dotted",
              data = filter(nycomps_roc,
                            metagroup == "genes",
                            group == "all",
                            outcome_threshold == "mix")) +
    geom_text(aes(label = count), x = .70, y = .19, size = 6*5/14, parse = TRUE,
              data = filter(p2_posmix_roc_label, metagroup == "genes", outcome_threshold == "mix")) +
    geom_text(aes(label = ci), x = .70, y = .10, size = 6*5/14, parse = TRUE,
              data = filter(p2_posmix_roc_label, metagroup == "genes", outcome_threshold == "mix")) +
    geom_text(aes(label = count), x = .70, y = .37, size = 6*5/14, parse = TRUE,
              data = filter(p2_posmix_roc_label, metagroup == "largescale")) +
    geom_text(aes(label = ci), x = .70, y = .28, size = 6*5/14, parse = TRUE,
              data = filter(p2_posmix_roc_label, metagroup == "largescale")) +
    # annotate("text", x = 25, y = 95, label = "By Gene:", size = 7*5/14, fontface = "bold") + 
    scale_x_continuous(expand = c(0, 0), breaks=seq(0, 1, .25), limits = c(0, 1), labels = function(x){x*100}) + 
    scale_y_continuous(expand = c(0, 0), breaks=seq(0, 1, .25), limits = c(0, 1), labels = function(x){x*100}) + 
    xlab("False Positive Rate") + 
    ylab("True Positive Rate") +
    scale_fill_manual(values = c("#A4A5A5","#A10F1C")) + 
    theme(aspect.ratio = 1, legend.position = "None", plot.margin=unit(c(0, 0, 0, 0), "in"))


prep_for_polygon <- function(df, x = "thresholds", y = "ppv", min_y = "min_ppv") {
    # prep_for_polygon(data.frame(thresholds = c(1, 2, -3), ppv = c(1, 5, 10)), min_ppv = 1)
    # order by thresholds
    df <- df[order(df[,x], decreasing = TRUE),]
    
    # min y value
    min_y_val <- df[, min_y][1]
        
    # top of polygon
    df$y_greater_min <- df[,y] > min_y_val
    df <- df[seq_len(nrow(df)) %>% rep(each=2) %>% tail(-1) %>% head(-1),]
    rownames(df) <- seq(length = nrow(df))
    df$id <-  rep(seq(nrow(df)/2), each = 2)
    df$type <- y
    
    # bottom of polygon
    df <- df[rep(seq_len(nrow(df)), each=2),]
    df[grepl(".1", rownames(df), fixed = TRUE),]$type <- min_y
    df[df$type == min_y, y] <- min_y_val
    
    # order for geom_polygon (needs to be closed)
    foreach(thisid = unique(df$id), .combine = rbind) %do% {
        thisdata <- filter(df, id == thisid) 
        thisdata <- thisdata[order(thisdata[, x], thisdata[,y], decreasing = TRUE),]
        rbind(thisdata[1,], thisdata[2,], thisdata[4,], thisdata[3,])
    }
}

p2_posmix_ppv_prep <- prep_for_polygon(filter(nycomps_roc, group == "posmix"))
p2_posmix_ppv <- ggplot(p2_posmix_ppv_prep) + 
    geom_polygon(aes(x=threshold_percentiles*100, y=ppv*100, group=id, fill=y_greater_min, color = y_greater_min), size = 0.1) +
    geom_path(aes(x=threshold_percentiles*100, y=ppv*100), size = 0.25, 
              data = filter(p2_posmix_ppv_prep, type != "min_ppv") %>% distinct(thresholds, ppv)) +
    geom_hline(aes(yintercept = min(min_ppv)*100), linetype="dashed", color="darkgrey", size = 0.5) +
#    annotate("text", x = 15, y = 55, label = "By Gene:", size = 7*5/14, fontface = "bold", family = "Helvetica") + 
    scale_x_continuous(expand=c(0,0), limits=c(0,100)) +
    scale_y_continuous(expand=c(0,0), limits=c(0,60), breaks=seq(0, 100, 20)) +
    xlab(expression("Percentile SVM"^"rank"*~"score")) + 
    ylab("PPV") +
    scale_color_manual(values = c("#A4A5A5","#A10F1C")) + 
    theme(legend.position="None", plot.margin=unit(c(1/16, 1/16, 1/16, 0), "in"))

p2_trial_counts <- ggplot(group_by(nycomps_vectors, id) %>% 
                              summarize(count = length(id), 
                                        pos_neg = factor(pos_neg[1], levels = c("neg", "mix", "pos")))) +
    geom_bar(aes(x = factor(count, levels = 1:8), fill = pos_neg), position = "stack", width = 0.5) +
    xlab("Number of Trials") +
    ylab("Number of Different Genes") + #AD665D
    scale_fill_manual(values = c("#A4A5A5", "#AD665D", "#A10F1C")) +
  #  scale_fill_manual(values = colorRampPalette(c("#A4A5A5", "#A10F1C"))(3)) +
    scale_y_continuous(expand=c(0,0)) +
    scale_x_discrete(expand=c(0,0), drop=FALSE) +
    theme(legend.position="none")

do_rollmean <- function(df, window) {
    df <- filter(df, abs(thresholds) != Inf)
    df <- data.frame(
        thresholds = zoo::rollmean(df$thresholds, window),
        ppv = zoo::rollmean(df$ppv, window)
        )
    df$window <- as.character(window)
    df
}

nycomps_nvector_ppv_smooth <- filter(nycomps_roc, group == "N") %>% select(thresholds, ppv) %>% arrange(desc(thresholds))
nycomps_nvector_ppv_smooth$window <- "1"
nycomps_nvector_ppv_smooth <- rbind(nycomps_nvector_ppv_smooth,
                                    do_rollmean(nycomps_nvector_ppv_smooth, 50),
                                    do_rollmean(nycomps_nvector_ppv_smooth, 100)
                                    )
nycomps_nvector_ppv_smooth$window <- factor(nycomps_nvector_ppv_smooth$window)

p4_nvector_ppv_prep <- prep_for_polygon(filter(nycomps_roc, group == "N"))

p4_nvector_ppv <- ggplot(p4_nvector_ppv_prep) + 
    geom_polygon(aes(x=thresholds, y=ppv*100, group=id, fill=y_greater_min, color = y_greater_min), size = 0.1) +
    geom_path(aes(x=thresholds, y=ppv*100), size = 0.5, data = filter(nycomps_nvector_ppv_smooth, window == "1")) +
#    geom_path(aes(x=thresholds, y=ppv*100), data = filter(nycomps_nvector_ppv_smooth, window == "50")) +
    geom_hline(aes(yintercept = min(min_ppv)*100), linetype="dashed", color="darkgrey", size = 0.5) +
    annotate("text", x = -1.7, y = 55, label = "Expression in (N) His-FLAG-TEV-", size = 7*5/14, fontface = "bold") + 
    coord_cartesian(xlim=nycomps_xscale) +
    scale_x_continuous(expand=c(0,0), breaks=nycomps_breaks) +
    scale_y_continuous(expand=c(0,0), limits=c(0,60), breaks=seq(0, 100, 20)) +
    xlab(expression("SVM"^"rank"*~"score")) + 
    ylab("Positive Predictive Value") +
    scale_color_manual(values = c("#A4A5A5", "#A10F1C")) + 
    scale_fill_manual(values = c("#A4A5A5", "#A10F1C")) + 
    theme(legend.position="None", plot.margin=unit(c(0, 0, 0, 0), "in"))

write.csv(nycomps_auc, file = "nycomps_auc.csv", row.names = FALSE)

nycomps_cterm <- filter(nycomps_auc, type == "plasmid" | (type == "all" & group == "vectors"), cterm == "all") %>%
    inner_join(filter(nycomps_auc, type == "plasmid" | (type == "all" & group == "vectors"), cterm != "all"), by = c("group" = "group")) %>%
    arrange(desc(count.x), cterm.y) %>%
    mutate(auc.x = specify_decimal(auc.x*100, k=1),
           lower95.x = specify_decimal(lower95.x*100, k=1),
           upper95.x = specify_decimal(upper95.x*100, k=1),
           auc.y = specify_decimal(auc.y*100, k=1),
           lower95.y = specify_decimal(lower95.y*100, k=1),
           upper95.y = specify_decimal(upper95.y*100, k=1)) %>%
    transmute(
        group = factor(group,
               levels = c("vectors", "N", "C", "C_LDAO", 
                          "MSGC.24", "MSGC.28", "MSGC.7", "MSGC.9", "MSGC.9_LDAO"),
               labels = c("All Conditions", "(N) His-FLAG-TEV-", "(C) -TEV-His", "(C) -TEV-His (LDAO)", 
                          "(24) His-GST-TEV-", "(28) -TEV-His", "(7) His-TEV-", "(9) His-MBP-TEV-", "(9) His-MBP-TEV- (LDAO)")),
        all_count = count.x,
        all_label = paste0(auc.x, " (", lower95.x, "-", upper95.x, ")"),
        cterm_count = count.y,
        cterm_label = paste0(auc.y, " (", lower95.y, "-", upper95.y, ")"),
        cterm = factor(cterm.y,
                       levels = c("CYTOPLASMIC.", "NON-CYTOPLASMIC."),
                       labels = c("Cyto.", "Peri.")))

nycomps_cterm <- filter(nycomps_cterm, group %in% c("All Conditions", "(N) His-FLAG-TEV-", "(C) -TEV-His"))

table_theme2 <-  ttheme_default(
    core = list(fg_params=list(fontsize = 7)),
    colhead = list(fg_params=list(fontsize = 7)),
    rowhead = list(fg_params=list(fontsize = 7)),
    padding = unit(c(1, 2), "mm"))

p2_auc_table <- tableGrob(select(nycomps_cterm, -group),
                          theme = table_theme2,
                          cols = c("Count", "AUC (95% CI)", "Count", "AUC (95% CI)", "cterm"),
                          rows = c("All Conditions", "", "(N) His-FLAG-TEV-", "", "(C) -TEV-His", ""))


# All Conditions
# ind <- find_cell(p2_auc_table, 1, 1, "rowhead-fg")
# p2_auc_table$layout[ind, c("t", "b", "z")] <- c(1, 2, zpos+1)
# ind <- find_cell(p2_auc_table, 1, 1, "rowhead-bg")
# p2_auc_table$layout[ind, c("t", "b", "z")] <- c(1, 2, zpos)

p2_auc_table <- change_cell(p2_auc_table, 2, 2, tnew = 2, bnew = 3)
p2_auc_table <- change_cell(p2_auc_table, 2, 3, tnew = 2, bnew = 3)

# N
p2_auc_table <- change_cell(p2_auc_table, 5, 2, tnew = 4, bnew = 5)
p2_auc_table <- change_cell(p2_auc_table, 5, 3, tnew = 4, bnew = 5)

# C
p2_auc_table <- change_cell(p2_auc_table, 6, 2, tnew = 6, bnew = 7)
p2_auc_table <- change_cell(p2_auc_table, 6, 3, tnew = 6, bnew = 7)

p2_main <- ggdraw() +
    draw_plot(
        plot_grid(p2_posmix_ppv, p2_posmix_histogram, 
                  labels = c("b", "c"), label_size = 8, ncol = 1, align = "v"), 
        x = 1/3, y = .355, width = 2/3, height = 1-.355) + 
#    annotate(geom="text", x = 0.55, y = 0.98, label = "933 complete successes (~11%)", size = 6*5/14, color = "#A10F1C") +
#    annotate(geom="text", x = 0.55, y = 0.95, label = "1189 mixed outcomes (~14%)", size = 6*5/14, color = "#A10F1C") +
#    annotate(geom="text", x = 0.55, y = 0.92, label = "6325 complete failures (~75%)", size = 6*5/14, color = "#A4A5A5") +
    draw_plot(p2_trial_counts, x = 0, y = .5, width = .33, height = .35) + 
    draw_plot(p2_posmix_roc, x = -.1, y = 0, width = .5, height = .35) + 
    draw_plot(p2_auc_table, x = .32, y = 0, width = 2/3, height = .35) + 
    draw_plot_label(label = c("a", "d", "e"), x = c(0, 0, 1/3), y = c(1, .35, .35), size = 8)
    
save_plot(filename = "plots/fig2_w_panels_largescale.pdf", plot = p2_main, base_width = uconv(136, "mm", "in", "Length")) #, base_height = 6)

```

# p3  tb, archaea, gpcrs, surade, and more!
# p3 ROCs (all)
```{r}
# archaea
archaea_scores <- read.csv("small-scale/archaea.fna.allstats.csv", header = TRUE, row.names=1)
rownames(archaea_scores) <- getname(rownames(archaea_scores))
archaea_scores$ml21 =  read.table("small-scale/archaea.fna.allstats.daley_only_raw_exp_ml21")$V1

archaea_outcomes <- read.csv("small-scale/archaea_S1.outcomes.csv", header = TRUE, stringsAsFactors = FALSE)
archaea_outcomes$ml21 <- archaea_scores[archaea_outcomes$Protein,]$ml21
rm(archaea_scores)

archaea_outcomes <- filter(archaea_outcomes, 
                           Measurement %in% c('WB/Gel', 'Scanning Densitometry of SDS-PAGE (%)'), 
                           !(Outcome %in% c("n.c.", "n.t.")))

# Key for Archaea Outcomes:
# "None", "WB-only", "1-10% of MP", ">10% of MP"
# "0", "1", "5", "10"

archaea_roc <- rbind(
    data.frame(
        my_roc(Outcome > 1 ~ ml21, direction = "<",
            data = filter(archaea_outcomes, Measurement != 'Scanning Densitometry of SDS-PAGE (%)')),
        positive_threshold = "sds_only"),
    data.frame(
        my_roc(Outcome > 0 ~ ml21, direction = "<",
            data = filter(archaea_outcomes, Measurement != 'Scanning Densitometry of SDS-PAGE (%)')),
        positive_threshold = "wb_sds")
)

archaea_auc <- rbind(
    data.frame(
        ci = ci.auc(Outcome > 1 ~ ml21, direction = "<", method = "delong",
            data = filter(archaea_outcomes, Measurement != 'Scanning Densitometry of SDS-PAGE (%)')),
        positive_threshold = "sds_only",
        ci_labels = c("lower95", "auc", "upper95"),
        pos_counts = sum(filter(archaea_outcomes, Measurement != 'Scanning Densitometry of SDS-PAGE (%)')$Outcome > 1),
        count = nrow(filter(archaea_outcomes, Measurement != 'Scanning Densitometry of SDS-PAGE (%)'))
        ),
    data.frame(
        ci = ci.auc(Outcome > 0 ~ ml21, direction = "<", method = "delong",
            data = filter(archaea_outcomes, Measurement != 'Scanning Densitometry of SDS-PAGE (%)')),
        positive_threshold = "wb_sds",
        ci_labels = c("lower95", "auc", "upper95"),
        pos_counts = sum(filter(archaea_outcomes, Measurement != 'Scanning Densitometry of SDS-PAGE (%)')$Outcome > 0),
        count = nrow(filter(archaea_outcomes, Measurement != 'Scanning Densitometry of SDS-PAGE (%)'))
        )
)
archaea_auc <- dcast(archaea_auc, positive_threshold + pos_counts + count ~ ci_labels, value.var = "ci")

archaea_median <- median(filter(archaea_outcomes, Measurement != 'Scanning Densitometry of SDS-PAGE (%)')$ml21)

p3_archaea_sdspage_data <- ddply(filter(archaea_outcomes, Measurement == 'Scanning Densitometry of SDS-PAGE (%)'),
                                .(Protein, ml21), summarize,
                                sds_max = max(as.numeric(Outcome)),
                                sds_min = min(as.numeric(Outcome)),
                                count = length(Outcome)
                                ) %>% filter(count != 1) 

p3_archaea_sdspage_data2 <- filter(archaea_outcomes, Measurement == 'Scanning Densitometry of SDS-PAGE (%)') %>% 
                   mutate(Outcome = as.numeric(Outcome)) %>%  
                   anti_join(p3_archaea_sdspage_data %>% gather(type, Outcome, sds_max:sds_min))

# Repeat of this one item, so spread them apart a tiny bit
p3_archaea_sdspage_data2 <- rbind(filter(p3_archaea_sdspage_data2, Protein != "Hbor39700" | Outcome != 12),
    filter(p3_archaea_sdspage_data2, Protein == "Hbor39700", Outcome == 12)[1,] %>% mutate(Outcome = 12.2),
    filter(p3_archaea_sdspage_data2, Protein == "Hbor39700", Outcome == 12)[2,] %>% mutate(Outcome = 11.8))

p3_archaea_sdspage <- ggplot(filter(archaea_outcomes, Measurement == 'Scanning Densitometry of SDS-PAGE (%)')) +
    geom_vline(xintercept = archaea_median-.015, linetype="dashed", color = "black", size = 0.5) +
    geom_point(aes(x=ml21, y=as.numeric(Outcome)), color="#A10F1C", size = 3, shape = 95, 
               data = p3_archaea_sdspage_data2) + 
    geom_errorbar(aes(x=ml21, ymin=sds_min, ymax=sds_max), color="#A10F1C", size = 0.27, width = 0.04,
                   data = p3_archaea_sdspage_data) +
    scale_y_continuous(expand = c(0,0), limits = c(0, 40)) +
    scale_x_continuous(expand = c(0,0), limits = c(-3.2, -1)) +
    ylab("Percentage of\nMembrane Protein") +
    xlab(expression("SVM"^"rank"*~"score")) + 
    scale_color_manual(values = brewer.pal(3, "Dark2")) +
    theme(axis.title = element_blank(),
              plot.margin = unit(c(1, 0, 2, 0), "mm"))

p3_archaea_western_data <- ddply(filter(archaea_outcomes, Measurement != 'Scanning Densitometry of SDS-PAGE (%)'),
                                   .(Protein), summarise,
                                   no = -sum(Outcome == 0),
                                   wb = sum(Outcome == 1),
                                   sdspage = sum(Outcome > 1),
                                   ml21 = ml21[[1]]
                                   ) %>% gather(group, count, no:sdspage) %>% filter(count != 0)
p3_archaea_western_data[p3_archaea_western_data$Protein == "MJ1319", "ml21"] <- 
    p3_archaea_western_data[p3_archaea_western_data$Protein == "MJ1319","ml21"] - 0.005
p3_archaea_western_data[p3_archaea_western_data$Protein == "Ta0252", "ml21"] <- 
    p3_archaea_western_data[p3_archaea_western_data$Protein == "Ta0252","ml21"] + 0.005

p3_archaea_western <- ggplot(p3_archaea_western_data) + 
    geom_vline(xintercept = archaea_median-.015, linetype="dashed", color = "black", size = 0.5) +
    geom_hline(yintercept = 0, alpha = 0.5, linetype="dashed", color = "darkgrey") +
    geom_bar(aes(x=ml21, y=count), fill="#A4A5A5",
             stat="identity", width=.02, data = filter(p3_archaea_western_data, group=="no", Protein == "MJ1319")) + 
    geom_bar(aes(x=ml21, y=count), fill="#A4A5A5",
             stat="identity", width=.02, data = filter(p3_archaea_western_data, group=="no", Protein != "MJ1319")) + 
    geom_bar(aes(x=ml21, y=count, fill=group), position = "stack",
             stat="identity", width=.02, data = filter(p3_archaea_western_data, group!="no")) +
    annotate("text", x = -2.9, y = 5.6, label = "Archaeal Transporters", size = 7*5/14, fontface = "bold") + 
    xlab(expression("SVM"^"rank"*~"score")) + 
    ylab("Count") +
    scale_fill_manual(values=c("#A10F1C", "#AD665D"), na.value="black") +
    scale_y_continuous(expand = c(0,0), limits = c(-6.1, 6), breaks=seq(-6, 6, 3)) +
    scale_x_continuous(expand = c(0,0), limits = c(-3.2, -1)) +
    theme(legend.position = "none",
              axis.title = element_blank(),
              plot.margin = unit(c(0, 0, 0, 0), "lines"))

archaea_auc_labels <- mutate(archaea_auc, 
           auc = specify_decimal(auc*100, k = 1),
           lower95 = specify_decimal(lower95*100, k = 1),
           upper95 = specify_decimal(upper95*100, k = 1),
#           ci = paste0("list(", pos_counts, ",~", auc, "~(", lower95, "-", upper95, "))"))
           ci = paste0(auc, "~(", lower95, "-", upper95, ")"))
archaea_auc_labels$y <- c(.16, .07)
archaea_counts_label <- paste0("list(", paste(archaea_auc$pos_counts %>% toString, sep = ",") %>% 
    paste(archaea_auc_labels$count[[1]], sep=" / "), ")")
p3_archaea_roc <- ggplot(archaea_roc) + 
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
    geom_path(aes(x=1-specificities, y=sensitivities, color = factor(positive_threshold))) +
    geom_text(aes(x = .65, y = y, label=ci, color=positive_threshold), size = 5*5/14, parse=TRUE, data=archaea_auc_labels) +
    annotate(geom="text", x = .65, y = .24, label = archaea_counts_label, size = 5*5/14, parse=TRUE) + 
    xlab("False Positive Rate") +
    ylab("True Positive Rate") +
    ggtitle("Archaeal Transporters") + 
    scale_color_manual(values = c("#A10F1C", "#AD665D"), na.value="black") +
    scale_y_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) + 
    scale_x_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) +
    theme(legend.position="None", aspect.ratio = 1,
              plot.margin = unit(c(0, 0, 0, 0), "lines"), 
              axis.title = element_text(size = 6))


### M. TUBERCULOSIS OUTCOMES AND SCORES (FROM TIM CROSS'S FIRST PAPER)
tcross_outcomes <- read.csv("small-scale/tcross_mt_outcomes.csv", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
tcross_outcomes <- mutate(tcross_outcomes, 
                          cloned = Cloned,
                          host = strain,
                          IB_CB = IB_CB*2,
                          SOL_CB = SOL_CB*2,
                          MEM_CB = MEM_CB*2,
                          IB_WB = as.numeric(IB_WB),
                          SOL_WB = as.numeric(SOL_WB),
                          MEM_WB = as.numeric(MEM_WB)
                          )

tcross_outcomes[is.na(tcross_outcomes$IB_WB),]$IB_WB  <- 0
tcross_outcomes[is.na(tcross_outcomes$SOL_WB),]$SOL_WB  <- 0
tcross_outcomes[is.na(tcross_outcomes$MEM_WB),]$MEM_WB  <- 0

tcross_outcomes$IB <- pmax(tcross_outcomes$IB_CB, tcross_outcomes$IB_WB)
tcross_outcomes$SOL <- pmax(tcross_outcomes$SOL_CB, tcross_outcomes$SOL_WB)
tcross_outcomes$MEM <- pmax(tcross_outcomes$MEM_CB, tcross_outcomes$MEM_WB)

tcross_outcomes <- select(tcross_outcomes, ORF, cloned, host, IB, SOL, MEM)

tcross_outcomes <- gather(tcross_outcomes, fraction, outcome, IB:MEM)
tcross_outcomes <- mutate(tcross_outcomes,
                          outcome = factor(outcome, levels=c(0,1,2), labels=c("No","WB","CB")),
                          fraction = factor(fraction, levels=c("MEM", "SOL", "IB"), labels=c("MEM", "SOL", "IB"))
                          )

tcross_scores <- data.frame(ml21 = read.table("small-scale/tcross.fna.allstats.daley_only_raw_exp_ml21", header = FALSE)$V1,
                            names = read.csv("small-scale/tcross.fna.allstats.csv", header = TRUE, stringsAsFactors = FALSE)$title)
tcross_scores$names <- getname(tcross_scores$names)

tcross_outcomes <- left_join(tcross_outcomes, tcross_scores, by = c("ORF" = "names"))
rm(tcross_scores)


tcross_C <- data.frame(ml21_vector = read.table("small-scale/tcross_C.fna.allstats.daley_only_raw_exp_ml21", 
                                                header = FALSE)$V1,
                       names = read.csv("small-scale/tcross_C.fna.allstats.csv", header = TRUE, stringsAsFactors = FALSE)$title)
tcross_C <- mutate(tcross_C, 
                   names = getname(names),
                   cloned = "C"
                   )
tcross_outcomes <- left_join(tcross_outcomes, tcross_C, by = c("ORF"="names", "cloned"="cloned"))
rm(tcross_C)

tcross_N <- data.frame(ml21_vector_N = read.table("small-scale/tcross_N.fna.allstats.daley_only_raw_exp_ml21", 
                                                header = FALSE)$V1,
                       names = read.csv("small-scale/tcross_N.fna.allstats.csv", header = TRUE)$title)
tcross_N <- mutate(tcross_N, 
                   names = getname(names),
                   cloned = "N"
                   )
tcross_outcomes <- left_join(tcross_outcomes, tcross_N, by = c("ORF"="names", "cloned"="cloned"))
rm(tcross_N)

tcross_outcomes[tcross_outcomes$cloned == 'N',]$ml21_vector <- tcross_outcomes[tcross_outcomes$cloned == 'N',]$ml21_vector_N
tcross_outcomes <- select(tcross_outcomes, -ml21_vector_N)

# > select(tcross_outcomes, ORF, cloned, host) %>% distinct %>% group_by(cloned, host) %>% summarize(tot_cloned = length(cloned))
# Source: local data frame [3 x 3]
# Groups: cloned [?]
# 
#   cloned              host tot_cloned
#    (chr)             (chr)      (int)
# 1      C CodonPlus-RP(DE3)         12
# 2      N          C42(DE3)          4
# 3      N CodonPlus-RP(DE3)         93

# > table(tcross_outcomes$outcome)
#  No  WB  CB 
# 208  66  53 
# > table(as.numeric(tcross_outcomes$outcome))
#   1   2   3 
# 208  66  53 
tcross_roc <- foreach(thisthreshold = c(2, 3), .combine='rbind') %dopar% {
    foreach(thisfraction=unique(tcross_outcomes$fraction), .combine='rbind') %dopar% {
        thisdata <- filter(tcross_outcomes, fraction == thisfraction)
        thisdata <- mutate(thisdata, thisresponse = as.numeric(outcome) >= thisthreshold)
        roc1 <- data.frame(
            my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
            fraction = thisfraction,
            positive_threshold = thisthreshold,
            score_type = "ml21")
        roc2 <- data.frame(
            my_roc(thisresponse ~ ml21_vector, data = thisdata, direction = "<"),
            fraction = thisfraction,
            positive_threshold = thisthreshold,
            score_type = "ml21_vector")
        rbind(roc1, roc2)
    }
}
tcross_roc$positive_threshold <- factor(tcross_roc$positive_threshold, 
                                        levels = c(2, 3), 
                                        labels = c("WB", "CB"))

tcross_auc <- foreach(thisthreshold = c(2, 3), .combine='rbind') %dopar% {
    foreach(thisfraction=unique(tcross_outcomes$fraction), .combine='rbind') %dopar% {
        thisdata <- filter(tcross_outcomes, fraction == thisfraction)
        thisdata <- mutate(thisdata, thisresponse = as.numeric(outcome) >= thisthreshold)
        data.frame(
            ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
            ml21_vector = ci.auc(thisresponse ~ ml21_vector, data = thisdata, direction = "<", method = "delong"),
            fraction = thisfraction,
            positive_threshold = thisthreshold,
            count = length(thisdata$thisresponse),
            pos_count = sum(thisdata$thisresponse),
            ci_labels = c("lower95", "auc", "upper95")
            )
        }
    }
tcross_auc$positive_threshold <- factor(tcross_auc$positive_threshold, 
                                        levels = c(2, 3), 
                                        labels = c("WB", "CB"))

tcross_auc <- gather(tcross_auc, score_type, ci, ml21:ml21_vector)
tcross_auc <- dcast(tcross_auc, fraction + positive_threshold + score_type + count + pos_count ~ ci_labels, value.var = "ci")

tcross_outcomes <- gather(tcross_outcomes, score_type, score_value, ml21:ml21_vector)

tcross_median <- ddply(tcross_outcomes, .(score_type, fraction), summarize,
                       median = median(score_value))

p3_mt <- ggplot(filter(tcross_outcomes, score_type == "ml21", fraction == "MEM")) +
    geom_vline(aes(xintercept = median(score_value)), linetype="dashed", size = 0.5) + 
    geom_jitter(aes(x=score_value, y=outcome, fill=as.numeric(outcome), color=as.numeric(outcome)) , shape=21, size = .95,
            position = position_quasirandom(varwidth = TRUE), 
            data=filter(tcross_outcomes, score_type == "ml21", fraction == "MEM")) + # host != "C42(DE3)",
#    geom_jitter(aes(x=score_value, y=outcome, fill=as.numeric(outcome)), color="black", shape=21, size = .95,
#                position = position_quasirandom(varwidth = TRUE), 
#                data=filter(tcross_outcomes, score_type == "ml21", host == "C42(DE3)", fraction == "MEM")) +
    annotate("text", x = -3.7, y = 3.1, label = "M. tuberculosis", size = 7*5/14, fontface = "bold.italic") + 
    scale_y_discrete(labels=c("WB" = "Western\nBlot", "CB" = "Coomassie\nBlue", "Neither" = "No\nExpression")) +
    scale_x_continuous(expand = c(.01,0), breaks = c(-5:5)) +
    scale_fill_gradient(low="#A4A5A5", high="#A10F1C", na.value="black") +
    scale_color_gradient(low="#A4A5A5", high="#A10F1C", na.value="black") +
    theme( #plot.title = element_text(face = "italic", size = 8),
              axis.title = element_blank(),
              legend.position="None",
              plot.margin = unit(c(0, 0, 0, 0), "lines"))

tcross_auc_labels <- filter(tcross_auc, fraction == "MEM", score_type == "ml21") %>%
    mutate(auc = specify_decimal(auc*100, k = 1),
           lower95 = specify_decimal(lower95*100, k = 1),
           upper95 = specify_decimal(upper95*100, k = 1),
           ci = paste0(auc, "~(", lower95, "-", upper95, ")"))
tcross_auc_labels$y <- c(.07, .16)
tcross_auc_counts <- tcross_auc_labels$pos_count %>% rev %>% toString %>% paste
tcross_auc_counts <- paste("list(", tcross_auc_counts, "/", tcross_auc_labels$count[[1]], ")")

p3_mt_roc <- ggplot(filter(tcross_roc, fraction == "MEM", score_type == "ml21")) + 
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
    geom_path(aes(x=1-specificities, y=sensitivities, color = factor(positive_threshold))) +
    geom_text(aes(x=.65, y=y, label=ci, color=positive_threshold), size = 5*5/14, parse = TRUE, data=tcross_auc_labels) +
    annotate(geom="text", x = .65, y = .25, label = tcross_auc_counts, size = 5*5/14, parse = TRUE) +
    scale_color_manual(values = ggplot_build(p3_mt)$data[[2]] %>% 
                           arrange(desc(-group)) %>% select(colour) %>% unique %>% unlist %>% unname %>% tail(-1), 
                         na.value="black") +
    ylab("True Positive Rate") +
    xlab("False Positive Rate") +
    ggtitle("M. tuberculosis") +
    scale_y_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) + 
    scale_x_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) +
    theme(legend.position="None",
              aspect.ratio = 1,
              plot.margin = unit(c(0, 0, 0, 0), "lines"),
              plot.title = element_text(size = 7, face = "italic"),
              axis.title = element_text(size = 6))


# GPCRs
lundstrom_scores <- read.csv("small-scale/lundstrom1.fna.allstats.csv", header = TRUE, row.names = 1)
lundstrom_scores$ml21 <- read.table("small-scale/lundstrom1.fna.allstats.daley_only_raw_exp_ml21")$V1

lundstrom_outcomes <- read.csv("small-scale/lundstrom1.060915.outcomes.csv", header = TRUE, stringsAsFactors = FALSE)
lundstrom_outcomes <- filter(lundstrom_outcomes, trust == 1)
rownames(lundstrom_outcomes) <- lundstrom_outcomes$new.ids

lundstrom_outcomes$ml21 <- lundstrom_scores[rownames(lundstrom_outcomes),]$ml21
rm(lundstrom_scores)

lundstrom_outcomes <- gather(lundstrom_outcomes, host, outcome, E..coli:SFV)
lundstrom_outcomes <- filter(lundstrom_outcomes, outcome != "nd")
lundstrom_outcomes$outcome <- factor(lundstrom_outcomes$outcome, 
                                     labels = c("No", "Low", "Medium", "High"),
                                     levels = c("0", "1", "2", "3"))

lundstrom_outcomes <- select(lundstrom_outcomes, new.ids, ml21, host, outcome)

lundstrom_median <- ddply(lundstrom_outcomes, .(host), summarize,
                          median = median(ml21))

lundstrom_roc <- foreach(thishost=unique(lundstrom_outcomes$host), .combine='rbind') %dopar% {
    data.frame(
        my_roc(outcome == "High" ~ ml21, data = filter(lundstrom_outcomes, host == thishost), direction = "<"),
        host = thishost)
    }

lundstrom_auc <- ddply(lundstrom_outcomes, .(host), summarize,
                     auc_geno = auc(response = outcome != "No", predictor = ml21, direction = "<"),
                     auc_onlyhigh = auc(response = outcome == "High", predictor = ml21, direction = "<")
                     )

# > table(lundstrom_outcomes$outcome)
#     No    Low Medium   High 
#     56     77     54     87 
# > table(as.numeric(lundstrom_outcomes$outcome))
#  1  2  3  4 
# 56 77 54 87 
lundstrom_roc <- foreach(thisthreshold = c(2, 3, 4), .combine='rbind') %dopar% {
    # ROC for each host separately
    data <- foreach(thishost=unique(lundstrom_outcomes$host), .combine='rbind') %dopar% {
        thisdata <- filter(lundstrom_outcomes, host == thishost)
        thisdata <- mutate(thisdata, thisresponse = as.numeric(outcome) >= thisthreshold)
        df <- data.frame(
            my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
            host = thishost,
            positive_threshold = thisthreshold)
        rbind(df, NA)
    }
   rbind(data, NA)
}

lundstrom_auc <- foreach(thisthreshold = c(2, 3, 4), .combine='rbind') %dopar% {
    # ROC for each host separately
    foreach(thishost=unique(lundstrom_outcomes$host), .combine='rbind') %dopar% {
        thisdata <- filter(lundstrom_outcomes, host == thishost)
        thisdata <- mutate(thisdata, thisresponse = as.numeric(outcome) >= thisthreshold)
        data.frame(
            ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
            host = thishost,
            positive_threshold = thisthreshold,
            count = length(thisdata$thisresponse),
            pos_count = sum(thisdata$thisresponse),
            ci_labels = c("lower95", "auc", "upper95")
          )
    }
}
lundstrom_auc <- dcast(lundstrom_auc, host + positive_threshold + count + pos_count ~ ci_labels, value.var = "ml21")

# We need to do it this way to break the x axis
p3_gpcr <- ggplot(filter(lundstrom_outcomes, host %in% c("E..coli", "P..pastoris"))) +
    geom_quasirandom(aes(x=host, y=ml21, color=outcome), dodge.width=.7, varwidth = TRUE, size = 1) +
    geom_segment(aes(y=median, yend=median, x=1.5, xend=2.5), size = 0.5, linetype = "dashed", 
                 data = filter(lundstrom_median, host == "E..coli")) +
    geom_segment(aes(y=median, yend=median, x=0.5, xend=1.5), size = 0.5, linetype = "dashed", 
                 data = filter(lundstrom_median, host == "P..pastoris")) +
    scale_color_manual(values = c('#A4A5A5', '#AC7B74', '#AA4F46', '#A10F1C')) +
    scale_x_discrete(expand=c(0,0), labels=c("P..pastoris" = "P. pastoris", "E..coli" = "E. coli")) +
    ylab(expression("SVM"^"rank"*~"score")) + 
    coord_flip()

p3_gpcr_sub_left <- p3_gpcr + 
    scale_y_continuous(expand=c(1,0), breaks = c(-10)) +
    coord_flip(ylim = c(-12,-10))  + 
    theme(legend.position="None",
              axis.title.x = element_blank(),
              axis.text.y = element_text(face="italic", color = brewer.pal(3, "Dark2")[c(2,1)]),
              axis.title.y = element_blank(),
              plot.margin = unit(c(2, 0, 2, 0), "mm"))

p3_gpcr_sub_right <- p3_gpcr + 
    annotate("text", y = -4.6, x = 2.36, label = "Mammalian GPCRs", size = 7*5/14, fontface = "bold") + 
    scale_y_continuous(limits=c(-5.4, 0.6), expand=c(0,0)) + 
    theme(legend.position="None", 
              axis.title.y = element_blank(),
              axis.text.y = element_blank(),
              axis.ticks.y = element_blank(),
              axis.line.y = element_blank(),
              plot.margin = unit(c(2, 0, 2, 0), "mm"))

# for some reason geom_quasirandom has to be added before geom_segment,
# so change the order to have the segments in back here:
p3_gpcr_sub_right <- ggplotGrob(p3_gpcr_sub_right)
p3_gpcr_sub_right$grobs[[4]]$childrenOrder <- p3_gpcr_sub_right$grobs[[4]]$childrenOrder[c(1,3,4,2,5)]

p3_gpcr_main <- plot_grid(p3_gpcr_sub_left, p3_gpcr_sub_right, align = "h", nrow=1, rel_widths = c(1.3, 10))

lundstrom_auc_labels <- filter(lundstrom_auc, host %in% c("E..coli", "P..pastoris")) %>% arrange(host, -positive_threshold) %>% 
    mutate(auc = specify_decimal(auc*100, k = 1),
           lower95 = specify_decimal(lower95*100, k = 1),
           upper95 = specify_decimal(upper95*100, k = 1),
           ci = paste0("'", auc, "'~(", lower95, "-'", upper95, "')")) 
lundstrom_auc_labels$y <- c(.25, .16, .07, .25, .16, .07)
lundstrom_auc_labels$x <- c(.65, .65, .65, .75, .75, .75)

lundstrom_auc_counts <- group_by(lundstrom_auc, host) %>% 
    summarise(pos_count = pos_count %>% rev %>% toString %>% paste, 
              count = count[[1]]) %>% 
    ungroup %>%
    mutate(label = paste0("list(", pos_count, "/", count, ")")) %>% 
    filter(host %in% c("E..coli", "P..pastoris"))
lundstrom_auc_counts$x <- c(.75, .65)

lundstrom_roc$host <- factor(lundstrom_roc$host,
                             levels = c("E..coli", "P..pastoris", "SFV"))

p3_gpcr_roc <- ggplot(filter(lundstrom_roc, host != "SFV")) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
    geom_path(aes(x=1-specificities, y=sensitivities, color = factor(positive_threshold))) +
    geom_text(aes(x=x, y=y, label=ci, color=factor(positive_threshold)), parse=TRUE, size = 5*5/14, data=lundstrom_auc_labels) +
    geom_text(aes(x=x, y=0.34, label=label), parse=TRUE, size=5*5/14, data=lundstrom_auc_counts) + 
    scale_color_manual(values = c( '#AC7B74', '#AA4F46', '#A10F1C')) +
    ylab("True Positive Rate") +
    xlab("False Positive Rate") +
    ggtitle("Mammalian GPCRs") + 
    scale_y_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) + 
    scale_x_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) +
    theme(legend.position="None",
              aspect.ratio=1,
              axis.title = element_text(size = 6),
              strip.text = element_text(size = 6, face = "italic", color=brewer.pal(3, "Dark2")[1:2]),
              plot.margin = unit(c(0, 0, 0, 0), "lines"),
              strip.switch.pad.wrap = unit(c(0, 0, 0, 0), "lines"),
              strip.switch.pad.grid = unit(c(0, 0, 0, 0), "lines")) +
    facet_wrap(~host, ncol=1, scales="free")

p3_gpcr_roc <- ggplotGrob(p3_gpcr_roc)
p3_gpcr_roc$grobs$strip_t1$grobs[[1]]$children[[2]]$children[[1]]$label <- "E. coli"
p3_gpcr_roc$grobs$strip_t2$grobs[[1]]$children[[2]]$children[[1]]$label <- "P. pastoris"
p3_gpcr_roc$grobs$strip_t2$grobs[[1]]$children[[2]]$children[[1]]$gp$col <- "#D95F02"

# e. coli Koth
koth_ec_scores <- read.csv("small-scale/koth_ec.ffn.allstats.csv_namefix", 
                           header = TRUE, stringsAsFactors = FALSE) %>% select(Gene)
koth_ec_scores$ml21 <- read.table("small-scale/koth_ec.ffn.allstats.daley_only_raw_exp_ml21")$V1

koth_ec_outcomes <- read.csv("small-scale/koth_ec.outcomes.csv", 
                             header = TRUE, stringsAsFactors = FALSE) %>%
    inner_join(koth_ec_scores, by = c("Gene" = "Gene"))

# remove these two since the calculated ORF is larger by 21 and 13 residues 
# than what's reported in the paper
# ycdG is greater by a single residue
koth_ec_outcomes <- filter(koth_ec_outcomes, !(Gene %in% c('yihO', 'kup')))

# t maritima Koth
tmaritima_scores <- read.csv("small-scale/koth_tmaritima.ffn.allstats.csv", header = TRUE, stringsAsFactors = FALSE) %>% 
    transmute(Gene = getname(title))
tmaritima_scores$ml21 <- read.table("small-scale/koth_tmaritima.ffn.allstats.daley_only_raw_exp_ml21")$V1

tmaritima_outcomes <- read.csv("small-scale/koth_tmaritima.outcomes.csv", header = TRUE, stringsAsFactors = FALSE) %>%
    select(-len.diff, -weight.diff) %>% 
    inner_join(tmaritima_scores, by = c("Gene" = "Gene"))

koth_ec_outcomes$group <- "E. coli"
tmaritima_outcomes$group <- "T. maritima"

koth_outcomes <- rbind(koth_ec_outcomes, tmaritima_outcomes) %>%
    mutate(group = factor(group),
           Purified = Purified * 2)
koth_outcomes$outcome <- factor(pmax(koth_outcomes$Expressed, koth_outcomes$Purified),
                                labels = c("Neither", "Expressed", "Purified"),
                                levels = c(0, 1, 2),
                                ordered = TRUE)

rm(koth_ec_scores, tmaritima_scores, koth_ec_outcomes, tmaritima_outcomes)


koth_roc <- foreach(thisgroup = unique(koth_outcomes$group), .combine = rbind) %dopar% {
    foreach(thisthreshold=levels(koth_outcomes$outcome) %>% tail(-1), .combine = rbind) %do% {
        thisdata <- filter(koth_outcomes, group == thisgroup)
        data.frame(
            group = thisgroup,
            positive_threshold = thisthreshold,
            my_roc(thisdata$outcome >= thisthreshold ~ ml21, direction = "<", data = thisdata)
            )
    }
}

koth_auc <- foreach(thisgroup = unique(koth_outcomes$group), .combine = rbind) %dopar% {
    foreach(thisthreshold=levels(koth_outcomes$outcome) %>% tail(-1), .combine = rbind) %do% {
        thisdata <- filter(koth_outcomes, group == thisgroup)
        data.frame(
            group = thisgroup,
            positive_threshold = thisthreshold,
            ci = ci.auc(outcome >= thisthreshold ~ ml21, direction = "<", data = thisdata),
            pos_count = sum(thisdata$outcome >= thisthreshold),
            count = length(thisdata$outcome),
            ci_labels = c("lower95", "auc", "upper95")
            )
    }
}

koth_auc <- dcast(koth_auc, group + positive_threshold + pos_count + count ~ ci_labels, value.var = "ci")

p3_koth <- ggplot(koth_outcomes) +
    geom_segment(aes(y=median(ml21), yend=median(ml21), x=1.5, xend=2.5), size = 0.5, linetype = "dashed", 
                 data = filter(koth_outcomes, group == "T. maritima")) +
    geom_segment(aes(y=median(ml21), yend=median(ml21), x=0.5, xend=1.5), size = 0.5, linetype = "dashed", 
                 data = filter(koth_outcomes, group == "E. coli")) +
    geom_quasirandom(aes(x=group, y=ml21, color=outcome), 
                     varwidth = TRUE, width = 0.3, dodge.width=.7, size=0.75) + 
    annotate("text", x = 2.36, y = -4.6, label = "E. coli & T. maritima", size = 7*5/14, fontface = "bold.italic") + 
    scale_color_manual(values=c("#A4A5A5","#AD665D","#A10F1C")) +
    scale_x_discrete(expand=c(0,0)) +
    ylab(expression("SVM"^"rank"*~"score")) + 
    theme(legend.position = "None",
              axis.title.y = element_blank(),
              axis.text.y = element_text(face= "italic")) +
    coord_flip()

p3_tmaritima <- ggplot(filter(koth_outcomes, group == "T. maritima")) +
    geom_vline(aes(xintercept=median(ml21)), linetype = "dashed") +
    geom_quasirandom(aes(x=ml21, y=factor(outcome), color=as.numeric(outcome)), varwidth = TRUE, width = 0.3, size=0.75) +
    annotate("text", x = -4.6, y = 3.36, label = "T. maritima", size = 7*5/14, fontface = "bold.italic") + 
    scale_color_gradient(low="#A4A5A5", high="#A10F1C", na.value="black") +
    xlab(expression("SVM"^"rank"*~"score")) + 
    theme(legend.position = "None",
              axis.title = element_blank())

koth_auc <- mutate(koth_auc, auc = specify_decimal(auc*100, k = 1),
                           lower95 = specify_decimal(lower95*100, k = 1),
                           upper95 = specify_decimal(upper95*100, k = 1),
                           ci = paste0("'", auc, "'~('", lower95, "'-'", upper95, "')")) %>%
    arrange(group, -as.numeric(positive_threshold))
koth_auc$y <- c(.16, .07)

koth_auc_counts <- group_by(koth_auc, group, count) %>%
    summarize(counts_label = pos_count %>% toString %>% paste) %>% ungroup %>%
    mutate(counts_label = paste0("list(", counts_label, "/", count, ")"))


p3_tmaritima_roc <- ggplot(filter(koth_roc, group == "T. maritima")) + 
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
    geom_path(aes(x=1-specificities, y=sensitivities, color = factor(positive_threshold))) +
    geom_text(aes(x=.65, y=y, label=ci, color=positive_threshold), parse=TRUE, size = 5*5/14, 
              data=filter(koth_auc, group == "T. maritima")) +
    annotate(geom="text", x=.65, y=.24, label=filter(koth_auc_counts, group == "T. maritima")$counts_label,
             size = 5*5/14, parse=TRUE) +
    scale_color_manual(values = c("#AD665D", "#A10F1C"), na.value="black") +
    ylab("True Positive Rate") +
    xlab("False Positive Rate") +
    ggtitle("T. maritima") +
    scale_y_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) + 
    scale_x_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) +
    theme(legend.position="None", aspect.ratio = 1,
              plot.margin = unit(c(0, 0, 0, 0), "lines"),
              plot.title = element_text(size = 7, face = "italic"),
        axis.title = element_text(size = 6))

p3_koth_roc <- ggplot(koth_roc) + 
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
    geom_path(aes(x=1-specificities, y=sensitivities, color = factor(positive_threshold))) +
    geom_text(aes(x=.65, y=y, label=ci, color=positive_threshold), parse=TRUE, size = 5*5/14, data=koth_auc) +
    geom_text(aes(x=.65, y=.24, label=counts_label), size = 5*5/14, parse=TRUE, data = koth_auc_counts) +
    scale_color_manual(values = c("#AD665D", "#A10F1C"), na.value="black") +
    ylab("True Positive Rate") +
    xlab("False Positive Rate") +
    ggtitle("Dobrovetsky, et al., 2005") +
    scale_y_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) + 
    scale_x_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) +
    facet_wrap(~group, ncol=1) +
    theme(legend.position="None", aspect.ratio = 1,
              plot.margin = unit(c(0, 0, 0, 0), "lines"),
        axis.title = element_text(size = 6))

# surade
surade_scores <- read.csv("small-scale/surade.fna.allstats.csv", header = TRUE, stringsAsFactors = FALSE) %>%
    transmute(
        gi = as.numeric(getname(title)),
        ml21 = read.table("small-scale/surade.fna.allstats.daley_only_raw_exp_ml21")$V1
    )

surade_outcomes <- read.csv("small-scale/surade.060915.outcomes.csv", header = TRUE, stringsAsFactors = FALSE) %>%
    inner_join(surade_scores, by = c("GI.number" = "gi")) %>%
    filter(!is.na(as.numeric(Outcome))) %>% 
    mutate(outcome = factor(Outcome, levels = c("0", "10", "100"), ordered = TRUE),
           vector = factor(Vector, 
                           labels = c("E. coli pTTQ18-A", "E. coli pTTQ18-C", "E. coli pQE-A", "E. coli pQE-C",
                                      "E. coli pBAD-A", "E. coli pBAD-C", "L. lactis pNZ-A", "L. lactis pNZ-C"),
                           levels = c("pTTQ18-A", "pTTQ18-C", "pQE-A", "pQE-C",
                                      "pBAD-A", "pBAD-C", "pNZ-A", "pNZ-C")),
           gi = GI.number) %>%
    select(vector, gi, outcome, ml21)

rm(surade_scores)

surade_roc <- foreach(thisthreshold = levels(surade_outcomes$outcome) %>% tail(-1), .combine='rbind') %dopar% {
    # ROC for each vector separately
    roc1 <- foreach(thisvector=levels(surade_outcomes$vector), .combine='rbind') %dopar% {
        thisdata <- filter(surade_outcomes, vector == thisvector) %>%
            mutate(thisresponse = outcome >= thisthreshold)
        data.frame(
            my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
            vector = thisvector,
            positive_threshold = thisthreshold)
    }
    # ROC for pET vectors together
    thisdata <- filter(surade_outcomes, vector %in%
                      c("E. coli pTTQ18-A", "E. coli pTTQ18-C", "E. coli pQE-A", "E. coli pQE-C"))
    thisdata <- mutate(thisdata, thisresponse = outcome >= thisthreshold)
    roc2 <- data.frame(
        my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
        vector = "pETonly",
        positive_threshold = thisthreshold)

    # ROC for pBAD together
    thisdata <- filter(surade_outcomes, vector %in% c("E. coli pBAD-A", "E. coli pBAD-C"))
    thisdata <- mutate(thisdata, thisresponse = outcome >= thisthreshold)
    roc3 <- data.frame(
        my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
        vector = "pBADonly",
        positive_threshold = thisthreshold)
    
    # ROC for all together
    thisdata <- surade_outcomes
    thisdata <- mutate(thisdata, thisresponse = outcome >= thisthreshold)
    roc4 <- data.frame(
        my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
        vector = "all",
        positive_threshold = thisthreshold)
    
    rbind(roc1, roc2, roc3, roc4)
}

surade_auc <- foreach(thisthreshold = levels(surade_outcomes$outcome) %>% tail(-1), .combine='rbind') %dopar% {
    # AUC for each vector separately
    auc1 <- foreach(thisvector=levels(surade_outcomes$vector), .combine='rbind') %dopar% {
        thisdata <- filter(surade_outcomes, vector == thisvector)
        thisdata <- mutate(thisdata, thisresponse = outcome >= thisthreshold)
        data.frame(
            ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
            vector = thisvector,
            positive_threshold = thisthreshold,
            pos_count = sum(thisdata$thisresponse),
            count = length(thisdata$thisresponse),
            ci_labels = c("lower95", "auc", "upper95")
          )
    }
    
    # AUC for pET vectors together
    thisdata <- filter(surade_outcomes, vector %in%
                      c("E. coli pTTQ18-A", "E. coli pTTQ18-C", "E. coli pQE-A", "E. coli pQE-C"))
    thisdata <- mutate(thisdata, thisresponse = outcome >= thisthreshold)
    auc2 <- data.frame(
        vector = "pETonly",
        ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method="delong"),
        positive_threshold = thisthreshold,
        pos_count = sum(thisdata$thisresponse),
        count = length(thisdata$thisresponse),
        ci_labels = c("lower95", "auc", "upper95")
        )

    # ROC for pBAD together
    thisdata <- filter(surade_outcomes, vector %in% c("E. coli pBAD-A", "E. coli pBAD-C"))
    thisdata <- mutate(thisdata, thisresponse = outcome >= thisthreshold)
    auc3 <- data.frame(
        vector = "pBADonly",
        ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method="delong"),
        positive_threshold = thisthreshold,
        pos_count = sum(thisdata$thisresponse),
        count = length(thisdata$thisresponse),
        ci_labels = c("lower95", "auc", "upper95")
        )
    
    # ROC for all E. coli together
    thisdata <- filter(surade_outcomes, vector %in% c("E. coli pTTQ18-A", "E. coli pTTQ18-C",
                                                      "E. coli pQE-A", "E. coli pQE-C",
                                                      "E. coli pBAD-A", "E. coli pBAD-C"))
    thisdata <- mutate(thisdata, thisresponse = outcome >= thisthreshold)
    auc4 <- data.frame(
        vector = "allec",
        ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method="delong"),
        positive_threshold = thisthreshold,
        pos_count = sum(thisdata$thisresponse),
        count = length(thisdata$thisresponse),
        ci_labels = c("lower95", "auc", "upper95")
        )

    # ROC for all L. lactis together
    thisdata <- filter(surade_outcomes, vector %in% c("L. lactis pNZ-A", "L. lactis pNZ-A"))
    thisdata <- mutate(thisdata, thisresponse = outcome >= thisthreshold)
    auc5 <- data.frame(
        vector = "all_ll",
        ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method="delong"),
        positive_threshold = thisthreshold,
        pos_count = sum(thisdata$thisresponse),
        count = length(thisdata$thisresponse),
        ci_labels = c("lower95", "auc", "upper95")
        )
    
    thisdata <- surade_outcomes
    thisdata <- mutate(thisdata, thisresponse = outcome >= thisthreshold)
    auc6 <- data.frame(
        vector = "all",
        ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method="delong"),
        positive_threshold = thisthreshold,
        pos_count = sum(thisdata$thisresponse),
        count = length(thisdata$thisresponse),
        ci_labels = c("lower95", "auc", "upper95")
        )
    
    rbind(auc1, auc2, auc3, auc4, auc5, auc6)
}

surade_auc <- dcast(surade_auc, vector + positive_threshold + pos_count + count ~ ci_labels, value.var = "ml21") 

surade_median <- ddply(surade_outcomes, .(vector), summarize,
                     median = median(ml21),
                     median_vector = median(ml21_vector)
                     )

p3_surade <- ggplot(filter(surade_outcomes, vector %in% 
                                    c("E. coli pTTQ18-A", "E. coli pTTQ18-C", "E. coli pQE-A", "E. coli pQE-C"))) +
    geom_vline(aes(xintercept=median(ml21)), linetype = "dashed") +
    geom_quasirandom(aes(x=ml21, y=outcome, color=as.numeric(outcome)), varwidth = TRUE, width = 0.3, size = 0.75) + 
    annotate("text", x = -2.7, y = 3.36, label = "Microbial Secondary Transporters", size = 7*5/14, fontface = "bold") + 
    scale_x_continuous(breaks=seq(-4, 1, 0.5)) +
    scale_color_gradient(low="#A4A5A5", high="#A10F1C", na.value="black") +
    xlab(expression("SVM"^"rank"*~"score")) + 
    ylab("Dot Density") +
    theme(legend.position = "None",
#              axis.title.x = element_blank(),
              axis.title.y = element_blank())

surade_auc_labels <- filter(surade_auc, vector == "pETonly") %>% 
    mutate(auc = specify_decimal(auc*100, k = 1),
           lower95 = specify_decimal(lower95*100, k = 1),
           upper95 = specify_decimal(upper95*100, k = 1),
           ci = paste0("'", auc, "'~(", lower95, "-", upper95, ")"),
           positive_threshold = factor(positive_threshold))
surade_auc_labels$y <- c(.07, .16)

surade_auc_counts <- filter(surade_auc, vector == "pETonly")$pos_count %>% rev %>% toString %>% paste
surade_auc_counts <- paste0("list(", surade_auc_counts,"/", filter(surade_auc, vector == "pETonly")$count[[1]], ")")

p3_surade_roc <- ggplot(filter(surade_roc, vector == "pETonly")) + 
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
    geom_path(aes(x=1-specificities, y=sensitivities, color = factor(positive_threshold))) +
    geom_text(aes(x=.65, y=y, label=ci, color=positive_threshold), parse=TRUE, size = 5*5/14, data=surade_auc_labels) +
    annotate(geom="text", x=.65, y=.24, label=surade_auc_counts, size = 5*5/14, parse=TRUE) +
    scale_color_manual(values = colorRampPalette(c("#A4A5A5", "#A10F1C"))(3)[2:3], na.value="black") +
    ylab("True Positive Rate") +
    xlab("False Positive Rate") +
    ggtitle("Microbial\nsecondary transporters") +
    scale_y_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) + 
    scale_x_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) +
    theme(legend.position="None", aspect.ratio = 1,
              plot.margin = unit(c(0, 0, 0, 0), "lines"),
        axis.title = element_text(size = 6))

# hpylori
hpylori_scores <- read.csv("small-scale/hpylori_psakis.ffn.allstats.csv", header = TRUE, stringsAsFactors = FALSE) %>%
    transmute(gene = title,
              ml21 = read.table("small-scale/hpylori_psakis.ffn.allstats.daley_only_raw_exp_ml21")$V1)

hpylori_outcomes <- read.csv("small-scale/hpylori-psakis.csv", header = TRUE, stringsAsFactors = FALSE) %>%
    filter(!is.na(as.numeric(Outcome))) %>% 
    mutate(gene = toupper(Code),
           outcome = factor(Outcome),
           group = factor(Vector,
                          levels = c("pET", "LAC / A", "LAC / M", "TEV / A", "TEV / M"),
                          labels = c("pET28 / pET36", "pQL60Lac / Automatic", "pQL60Lac / Manual", 
                                     "pQTev / Automatic", "pQTev / Manual")),
           vector = factor(substring(Vector, 1, 3)),
           scoring = factor(substrRight(Vector, 1))) %>%
    select(gene, group, vector, scoring, outcome) %>%
    inner_join(hpylori_scores, by = c("gene" = "gene"))

rm(hpylori_scores)

hpylori_outcomes_means <- group_by(hpylori_outcomes, gene, vector) %>%
    summarize(outcome = mean(as.numeric(as.character(outcome)), na.rm = TRUE),
              ml21 = ml21[[1]])

hpylori_median <- rbind(
    group_by(hpylori_outcomes, group) %>% summarise(
          count = length(ml21),
          median = median(ml21),
          type = "direct"
          ),
    data.frame(group = "all", 
           count = length(rbind(hpylori_outcomes, filter(hpylori_outcomes, vector == "pET"))$ml21),
           median = median(rbind(hpylori_outcomes, filter(hpylori_outcomes, vector == "pET"))$ml21),
           type = "direct"),
    group_by(hpylori_outcomes, vector) %>% summarise(
          count = length(ml21),
          median = median(ml21),
          type = "mean"
          ) %>% rename(group = vector),
    data.frame(group = "all", 
               count = length(hpylori_outcomes_means$ml21),
               median = median(hpylori_outcomes_means$ml21),
               type = "mean")
    )

hpylori_roc <- foreach(thisthreshold = c(1, 2, 3), .combine='rbind') %dopar% {
    # ROC for each vector separately
    roc1 <- foreach(thisgroup=unique(hpylori_outcomes$group), .combine='rbind') %dopar% {
        thisdata <- filter(hpylori_outcomes, group == thisgroup)
        thisdata <- mutate(thisdata, thisresponse = as.numeric(as.character(outcome)) >= thisthreshold)
        data.frame(
            my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
            group = thisgroup,
            positive_threshold = thisthreshold,
            type = "direct")
    }
    # ROC for all vectors together
    # need to double pET measurements since LAC and TEV have both manual and automated scores
    thisdata <- rbind(hpylori_outcomes, filter(hpylori_outcomes, vector == "pET"))
    thisdata <- mutate(thisdata, thisresponse = as.numeric(as.character(outcome)) >= thisthreshold)
    roc2 <- data.frame(
        my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
        group = "all",
        positive_threshold = thisthreshold,
        type = "direct")
    
    roc3 <- foreach(thisthreshold=c(thisthreshold, thisthreshold-0.5), .combine='rbind') %dopar% { 
        roc4 <- foreach(thisgroup=unique(hpylori_outcomes_means$vector), .combine='rbind') %dopar% {
            thisdata <- filter(hpylori_outcomes_means, vector == thisgroup)
            thisdata <- mutate(thisdata, thisresponse = as.numeric(as.character(outcome)) >= thisthreshold)
            data.frame(
                my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
                group = thisgroup,
                positive_threshold = thisthreshold,
                type = "mean")
        }
        
        # ROC for all vectors together (means)
        thisdata <- hpylori_outcomes_means
        thisdata <- mutate(thisdata, thisresponse = as.numeric(as.character(outcome)) >= thisthreshold)
        roc5 <- data.frame(
            my_roc(thisresponse ~ ml21, data = thisdata, direction = "<"),
            group = "all",
            positive_threshold = thisthreshold,
            type = "mean")
        
        rbind(roc4,roc5)
    }
    
    rbind(roc1, roc2, roc3)
}

hpylori_auc <- foreach(thisthreshold = c(1, 2, 3), .combine='rbind') %dopar% {
    # AUC for each vector separately
    auc1 <- foreach(thisgroup=unique(hpylori_outcomes$group), .combine='rbind') %dopar% {
        thisdata <- filter(hpylori_outcomes, group == thisgroup)
        thisdata <- mutate(thisdata, thisresponse = as.numeric(as.character(outcome)) >= thisthreshold)
        data.frame(
            ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
            group = thisgroup,
            positive_threshold = thisthreshold,
            type = "direct",
            count = length(thisdata$thisresponse),
            pos_count = sum(thisdata$thisresponse),
            ci_labels = c("lower95", "auc", "upper95")
          )
    }
    
    # AUC for all vectors together
    # need to double pET since only one measurement vs two ("Auto" and "Manual" for LAC and TEV)
    thisdata <- rbind(hpylori_outcomes, filter(hpylori_outcomes, vector == "pET"))
    thisdata <- mutate(thisdata, thisresponse = as.numeric(as.character(outcome)) >= thisthreshold)
    auc2 <- data.frame(
        group = "all",
        ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method="delong"),
        positive_threshold = thisthreshold,
        type = "direct",
        count = length(thisdata$thisresponse),
        pos_count = sum(thisdata$thisresponse),
        ci_labels = c("lower95", "auc", "upper95")
        )
    
    auc3 <- foreach(thisthreshold=c(thisthreshold, thisthreshold-0.5), .combine='rbind') %dopar% { 
        auc4 <- foreach(thisgroup=unique(hpylori_outcomes_means$vector), .combine='rbind') %dopar% {
            thisdata <- filter(hpylori_outcomes_means, vector == thisgroup)
            thisdata <- mutate(thisdata, thisresponse = as.numeric(as.character(outcome)) >= thisthreshold)
            data.frame(
                ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
                group = thisgroup,
                positive_threshold = thisthreshold,
                type = "mean",
                count = length(thisdata$thisresponse),
                pos_count = sum(thisdata$thisresponse),
                ci_labels = c("lower95", "auc", "upper95")
              )
        }
        
        # AUC for means outcomes of all vectors together
        thisdata <- hpylori_outcomes_means
        thisdata <- mutate(thisdata, thisresponse = as.numeric(as.character(outcome)) >= thisthreshold)
        auc5 <- data.frame(
            group = "all",
            ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method="delong"),
            positive_threshold = thisthreshold,
            type = "mean",
            count = length(thisdata$thisresponse),
            pos_count = sum(thisdata$thisresponse),
            ci_labels = c("lower95", "auc", "upper95")
            )
        
        rbind(auc4, auc5)
    }
    
    rbind(auc1, auc2, auc3)
}
hpylori_auc <- dcast(hpylori_auc, group + positive_threshold + type + count + pos_count ~ ci_labels, value.var = "ml21")

p3_hpylori <- ggplot(hpylori_outcomes_means) + 
    geom_vline(aes(xintercept=median(ml21)), size = 0.5, linetype = "dashed") +
    geom_point(aes(x=ml21, y=outcome, color=outcome), size=0.75, position=position_quasirandom(varwidth = TRUE)) + # diamond is shape=18
    annotate("text", x = -6.5, y = 3.25, label = "H. pylori", size = 7*5/14, fontface = "bold.italic") + 
    scale_color_gradient(low="#A4A5A5", high="#A10F1C", na.value="black") +
    scale_x_continuous(breaks=c(-8:1), expand=c(0.01, 0.01)) +
    scale_y_continuous(breaks=seq(0, 3, 0.5), expand=c(0.02, 0.02)) +
    xlab(expression("SVM"^"rank"*~"score")) + 
    ylab("Reported Outcome") +
    theme(legend.position="None",
              axis.title.y = element_blank(),
              axis.title.x = element_blank(),
              plot.margin = unit(c(2, 0, 0, 0), "mm"))

p3_hpylori_roc <- ggplot(filter(hpylori_roc, group == "all", type == "mean")) + 
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
    geom_path(aes(x=1-specificities, y=sensitivities, color = factor(positive_threshold)), size = 0.25) +
    scale_color_manual(values = colorRampPalette(c("#A4A5A5", "#A10F1C"))(7)[2:7], na.value="black") +
    ylab("True Positive Rate") +
    xlab("False Positive Rate") +
    ggtitle("H. pylori") + 
    scale_y_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) + 
    scale_x_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) +
    theme(legend.position="None", aspect.ratio = 1,
              plot.margin = unit(c(0, 0, 0, 0), "lines"),
              plot.title = element_text(size = 7, face = "italic"),
              axis.title = element_text(size = 6))

hpylori_auc_labels <- filter(hpylori_auc, group == "all", type == "mean") %>% arrange(-positive_threshold) %>%
    mutate(auc = specify_decimal(auc*100, k = 1),
           lower95 = specify_decimal(lower95*100, k = 1),
           upper95 = specify_decimal(upper95*100, k = 1),
           ci = paste0("'", auc, "'~('", lower95, "'-'", upper95, "')"))
hpylori_auc_labels$x <- rep(.5, 6)
hpylori_auc_labels$y <- c(.52, .43, .34, .25, .16, .07)
hpylori_auc_labels$y <- c(.60, .50, .40, .30, .20, .10)


hpylori_auc_counts <- filter(hpylori_auc, group == "all", type == "mean")$pos_count %>% rev %>% toString %>% paste
hpylori_auc_counts <- paste0("list(", hpylori_auc_counts, "/", filter(hpylori_auc, group == "all", type == "mean")$count[[1]], ")")

p3_hpylori_auc <- ggplot(hpylori_auc_labels) + 
    geom_text(aes(x=x, y=y, label=ci, color=factor(positive_threshold)), parse=TRUE, size = 5*5/14) +
    annotate(geom="text", x=.5, y=.70, label=hpylori_auc_counts, size = 5*5/14, parse=TRUE) +
    scale_color_manual(values = colorRampPalette(c("#A4A5A5", "#A10F1C"))(7)[2:7], na.value="black") +
#    scale_y_continuous(expand = c(0, 0), breaks=c(0, .5, 1), limits=c(0, 1), labels=function(x){x*100}) + 
    scale_x_continuous(expand = c(0, 0), limits=c(0, 1)) +
    theme(
        legend.position="None",
        plot.margin = unit(c(0, 0, 0, 0), "lines"),
        plot.title = element_blank(),
        axis.title = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        axis.text = element_blank())


p3_main <- ggdraw() +
    draw_plot(
        plot_grid(p3_archaea_western, p3_archaea_sdspage, p3_mt, NULL,
                  labels = c("a", "","","e"), label_size = 8, ncol = 1, align = "v", rel_heights = c(1.8, 1.8, 2.2, 3.2)),
        x = 0, y = 0, width = .8, height = 1) + 
    draw_plot_label("c", x = 0, y = .62, size = 8) +
    draw_plot(p3_gpcr_main, x = 0.003, y = 0, width = .797, height = .37) +
    annotate("segment", x = .0975, xend = .0935, y = .063, yend = .072) + 
    annotate("segment", x = .1010, xend = .0970, y = .063, yend = .072) + 
    draw_label("Number of Conditions", x = .03, y = .912, angle = 90, size = 7) + 
    draw_label("Percentage of\nMembrane Protein", x = .03, y = .71, angle = 90, size = 7) + 
    draw_label("Coomassie Blue", x = .22, y = .945, size = 6, colour = "#A10F1C") + 
    draw_label("Western Blot", x = .22, y = .925, size = 6, colour = "#AD665D") + 
    draw_label("No Expression", x = .22, y = .895, size = 6, colour = "#A4A5A5") + 
    draw_plot(
        plot_grid(p3_archaea_roc, p3_mt_roc, NULL,
                  labels = c("b","d","f"), label_size = 8, ncol = 1, align = "hv"),
        x = .82, y = .25, width = .16, height = .75) +
    draw_plot(p3_gpcr_roc, x = .815, y = 0, width = .17, height = .5)

save_plot(filename = "plots/fig3_w_panels.pdf", plot = p3_main, base_width = uconv(183, "mm", "in", "Length"), base_height = 5)

p3_extended <- ggdraw() +
    draw_plot(
        plot_grid(p3_hpylori, p3_tmaritima, p3_surade,
                  labels = c("a","c","e"), label_size = 8, ncol = 1, align = "v", rel_heights = c(5, 2.5, 2.5)),
        x = 0, y = 0, width = .8, height = 1) +
    draw_label("Expression Score", x = .03, y = .755, angle = 90, size = 7) + 
    draw_label("Dot Density", x = .03, y = .15, angle = 90, size = 7) + 
    draw_plot(
        plot_grid(p3_hpylori_roc, p3_hpylori_auc, p3_tmaritima_roc, p3_surade_roc, rel_heights = c(1, .8, 1, 1),
                  labels = c("b", "", "d", "f"), label_size = 8, ncol = 1, align = "hv"),
        x = .80, y = 0, width = .17, height = 1)
save_plot(filename = "plots/extfig1.pdf", plot = p3_extended, base_width = uconv(183, "mm", "in", "Length"), base_height = 5)

p3_koth
p3_koth_roc

p3_extended_koth_all <- ggdraw() +
    draw_plot(
        plot_grid(p3_hpylori, p3_surade, p3_koth,
                  labels = c("a","c","e"), label_size = 8, ncol = 1, align = "v", rel_heights = c(5, 2.5, 5)),
        x = 0, y = 0, width = .8, height = 1) +
    draw_label("Expression Score", x = .03, y = .755, angle = 90, size = 7) + 
    draw_label("Dot Density", x = .03, y = .385, angle = 90, size = 7) + 
    draw_plot(
        plot_grid(p3_hpylori_roc, p3_hpylori_auc, p3_surade_roc, p3_koth_roc, rel_heights = c(1, .8, 1, 2),
                  labels = c("b", "", "d", "f"), label_size = 8, ncol = 1, align = "hv"),
        x = .80, y = 0, width = .17, height = 1)
save_plot(filename = "plots/extfig1_koth_all.pdf", plot = p3_extended_koth_all, base_width = uconv(183, "mm", "in", "Length"), base_height = 7)
```

# p4 - forward predictions
```{r}
# load allstats and scores
celegans <- read.csv("forward-preds/celegans_mp_wormbase.fna.allstats.csv", header = TRUE, as.is = TRUE) %>%
    mutate(group="Metazoa",
           species="Caenorhabditis elegans",
           ml21 = read.table("forward-preds/celegans_mp_wormbase.fna.allstats.daley_only_raw_exp_ml21")$V1)

hs <- read.csv("forward-preds/hs_rnd1.fna.allstats.csv", header = TRUE, as.is = TRUE) %>%
    mutate(group="Metazoa",
           species="Homo sapiens",
           ml21 = read.table("forward-preds/hs_rnd1.fna.allstats.daley_only_raw_exp_ml21")$V1)

mm <- read.csv("forward-preds/mm_rnd1.fna.allstats.csv", header = TRUE, as.is = TRUE) %>%
    mutate(group="Metazoa",
           species="Mus musculus",
           ml21 = read.table("forward-preds/mm_rnd1.fna.allstats.daley_only_raw_exp_ml21")$V1)

sgd <- read.csv("forward-preds/sgd_orf_coding.mps.fna.allstats.csv", header = TRUE, as.is = TRUE) %>%
    mutate(group = "Eukaryota",
           species = "Saccharomyces cerevisiae",
           ml21 = read.table("forward-preds/sgd_orf_coding.mps.fna.allstats.ml21")$V1)

pipca <- read.csv("forward-preds/Picpa1_GeneCatalog_CDS_20130227.mps.fna.allstats.csv", header = TRUE, as.is = TRUE) %>%
    mutate(group = "Eukaryota",
           species = "Pichia pastoris",
           ml21 = read.table("forward-preds/Picpa1_GeneCatalog_CDS_20130227.mps.fna.allstats.ml21")$V1)

microbial_preds <- read.csv("forward-preds/microbial_query.fna.allstats.csv", header = TRUE, as.is = TRUE) %>%
    mutate(ml21 = read.table("forward-preds/microbial_query.fna.allstats.daley_only_raw_exp_ml21")$V1) %>%
    cbind(read.table("forward-preds/microbial_query.dat", header = TRUE, sep = "\t")) %>%
    select(-CONVERT..pfamseq.sequence.USING.UTF8., -sequence) %>%
    mutate(species = gsub("/", "|", species),
           species = getname(gsub(" ", "_", species)),
           species = gsub("_", " ", species)) %>%
    rename(group = grouping, uniprotid = pfamseq_acc) %>%
    filter(species != "Escherichia coli (strain ATCC 33849 ")

forward_preds <- rbind(celegans, mm, hs, sgd, pipca, microbial_preds[, colnames(pipca)]) %>%
    filter(numTMs > 0) %>%
    mutate(species = regmatches(species, regexpr('\\w+\\W{1,2}\\w+', species, perl=TRUE)))

# Two E. coli datasets here for plotting comparison
daley_outcomes <- read.csv("training/AllPlates_161104_with_raw.csv", header = TRUE) %>%
    filter(Cterm_new == "in", Keep. == 1) %>%
    transmute(gfp = Normalised.GFP..I.e..x.3924.,
              ID = tolower(ID)) %>%
    left_join(ecoli_training, by = c("ID" = "ID")) %>%
    dplyr::select(gfp, ID, ml21) %>%
    filter(gfp > 0)
daley_outcomes <- daley_outcomes[!duplicated(daley_outcomes),]
rownames(daley_outcomes) <- daley_outcomes$ID


fluman_outcomes <- read.csv("training/fluman_data_all.csv", header = TRUE)
rownames(fluman_outcomes) <- fluman_outcomes$name
fluman_outcomes <- select(fluman_outcomes, folded.exp, folded.exp.1, folded.exp.2)
fluman_outcomes <- fluman_outcomes/c(fluman_outcomes['emrD',])
fluman_outcomes$avg_folded <- 
    rowMeans(data.frame(fluman_outcomes$folded.exp, 
                        fluman_outcomes$folded.exp.1, 
                        fluman_outcomes$folded.exp.2), na.rm = TRUE)
fluman_outcomes$ID <- tolower(rownames(fluman_outcomes))
fluman_outcomes <- left_join(fluman_outcomes, ecoli_training, by = c("ID" = "ID")) %>%
    select(ID, avg_folded, ml21)
fluman_outcomes <- fluman_outcomes[!duplicated(fluman_outcomes),]
rownames(fluman_outcomes) <- fluman_outcomes$ID

# give a table with counts in each category, counts above score 0, counts above score 2
forward_tufte_boxplots <- filter(forward_preds, numTMs >= 1) %>% 
    group_by(group, species) %>%
    summarise(
      count = length(ml21),
      # for tufte like boxplots, if the min is < xrange min, then the whole linepart doesnt show:
      boxplot_1 = boxplot.stats(ml21, coef = 1.5, do.conf = TRUE, do.out = TRUE)$stats[[1]],
      boxplot_1 = ifelse(boxplot_1 < nycomps_xscale[[1]], nycomps_xscale[[1]], boxplot_1),
      boxplot_2 = boxplot.stats(ml21, coef = 1.5, do.conf = TRUE, do.out = TRUE)$stats[[2]],
      boxplot_3_median = boxplot.stats(ml21, coef = 1.5, do.conf = TRUE, do.out = TRUE)$stats[[3]],
      boxplot_4 = boxplot.stats(ml21, coef = 1.5, do.conf = TRUE, do.out = TRUE)$stats[[4]],
      boxplot_5 = boxplot.stats(ml21, coef = 1.5, do.conf = TRUE, do.out = TRUE)$stats[[5]],
      boxplot_5 = ifelse(boxplot_5 > nycomps_xscale[[2]], nycomps_xscale[[2]], boxplot_5)) %>%
    mutate(species_count = paste0(species, " (", count, ")") ) %>%
    arrange(species == "Escherichia coli", group == "Metazoa", boxplot_3_median) %>% ungroup

forward_ec_thresholds <- filter(forward_tufte_boxplots, species == "Escherichia coli") %>% 
    select(-group, -species, -count, -species_count) %>% matrix %>% unlist

forward_targets_table <- group_by(forward_preds, group, species) %>%
    summarise( 
        totalMPs = length(ml21),
        MPs_ge_ECmed = sum(ml21 >= forward_ec_thresholds[[3]]),
        MPs_ge_ECq3 = sum(ml21 > forward_ec_thresholds[[4]]),
        polyMPs_ge_ECmed = sum(ml21 > forward_ec_thresholds[[3]] & numTMs > 1),
        polyMPs_ge_ECq3 = sum(ml21 > forward_ec_thresholds[[4]] & numTMs > 1)
      ) %>% ungroup

forward_preds_order <- c(
    "Vibrio cholerae",
    "Helicobacter pylori",
    "Campylobacter jejuni",
    "Mycobacterium tuberculosis",
    "Staphylococcus aureus",
    "Plasmodium falciparum",
    
    "Aquifex aeolicus",
    "Thermus thermophilus",
    "Spirochaeta thermophila",
    
    "Gloeobacter violaceus",
    "Rhizobium loti",
    "Methanococcus maripaludis",
    "Haloarcula marismortui",
    "Bacillus halodurans",
    
    "Lactobacillus casei",
    "Bacillus cereus",
    "Bacillus subtilis",
    "Escherichia coli",
    
    "Pichia pastoris",
    "Saccharomyces cerevisiae",
    
    "Caenorhabditis elegans",
    "Mus musculus",           
    "Homo sapiens")

p4_all_violin <- ggplot(forward_preds) + 
    geom_hline(yintercept=forward_ec_thresholds[[3]], linetype="dashed", color="darkgrey") +
    geom_point(aes(x=species, y=boxplot_3_median), color = "darkgrey", data = forward_tufte_boxplots) + 
    geom_text(aes(x=species, y=-6, label = species_count), vjust = 2, fontface = "italic", size = 7*5/14, data = forward_tufte_boxplots) + 
    geom_linerange(aes(x=species, ymin=boxplot_1, ymax=boxplot_2), color = "darkgrey", data = forward_tufte_boxplots)+ 
    geom_linerange(aes(x=species, ymin=boxplot_4, ymax=boxplot_5), color = "darkgrey", data = forward_tufte_boxplots) +
    geom_violin2(aes(y=ml21, x=species), color = "black", adjust = 0.3, alpha = 0) + 
    # to get the tufte-like boxplots
    #geom_hline(yintercept=forward_ec_thresholds[[4]], linetype="dashed", color="darkgrey") +
    coord_flip() +
    ylab(expression("SVM"^"rank"*~"score")) + 
    scale_y_continuous(expand=c(0,0), limits=nycomps_xscale, breaks=nycomps_breaks) +
    scale_x_discrete(limits=forward_preds_order) +
    theme(axis.ticks.y = element_blank(),
              axis.title.y = element_blank(),
              axis.text.y = element_blank(),
              axis.line.y = element_blank())

selected_violin_genomes <- c("Plasmodium falciparum",
                    "Aquifex aeolicus",
                    "Thermus thermophilus",
                    "Spirochaeta thermophila",
                    "Bacillus subtilis",
                    "Escherichia coli",
                    "Saccharomyces cerevisiae",
                    "Homo sapiens")

p4_selected_violin <- ggplot(forward_preds %>% filter(species %in% selected_violin_genomes)) + 
    geom_hline(yintercept = forward_ec_thresholds[[3]], linetype = "dashed", color = "darkgrey") +
    geom_point(aes(x = species, y=boxplot_3_median), color = "darkgrey",
               data = forward_tufte_boxplots %>% filter(species %in% selected_violin_genomes)) + 
    geom_text(aes(x = species, y=-6, label = species_count), vjust = 2, fontface = "italic", size = 7*5/14,
              data = forward_tufte_boxplots %>% filter(species %in% selected_violin_genomes)) + 
    geom_linerange(aes(x = species, ymin = boxplot_1, ymax = boxplot_2), color = "darkgrey",
                   data = forward_tufte_boxplots %>% filter(species %in% selected_violin_genomes))+ 
    geom_linerange(aes(x = species, ymin = boxplot_4, ymax = boxplot_5), color = "darkgrey",
                   data = forward_tufte_boxplots %>% filter(species %in% selected_violin_genomes)) +
    geom_violin2(aes(y = ml21, x = species), color = "black", adjust = 0.3, alpha = 0) + 
    coord_flip() +
    ylab(expression("SVM"^"rank"*~"score")) + 
    scale_x_discrete(limits = selected_violin_genomes, breaks = selected_violin_genomes) +
    scale_y_continuous(expand = c(0,0), limits = nycomps_xscale, breaks = nycomps_breaks) +
    theme(axis.ticks.y = element_blank(),
              axis.title.y = element_blank(),
              axis.text.y = element_blank(),
              axis.line.y = element_blank())
```

## Load Fluman, et al., aSD mutant data (Figure 6 in their paper)
```{r}
rel_diff <- function(mut, wt, func = range) {
    2*(mut - wt)/abs((wt + mut))
}

fluman_asd <- read.csv("forward-preds/fig6-analysis.csv", header = TRUE, as.is = TRUE) %>%
    filter(Band == "FOLDED") %>%
    rename(outcome = Area.from.Image,
           ml21 = ml21.scores) %>%
    select(Type, Construct, outcome, ml21)

fluman_asd <- fluman_asd %>%
    filter(Construct == "WT") %>%
    inner_join(fluman_asd %>% filter(Construct == "MUT"), by = c("Type" = "Type")) %>%
    transmute(gene = toupper(Type),
              wt_outcome = outcome.x,
              wt_ml21 = ml21.x,
              mut_outcome = outcome.y,
              mut_ml21 = ml21.y,
              rel_outcome = rel_diff(mut_outcome, wt_outcome),
              rel_ml21 = rel_diff(mut_ml21, wt_ml21))

fluman_asd_features <- "forward-preds/fluman_asd_mutant.fna.allstats.csv" %>%
    read.csv(header = TRUE, as.is = TRUE) %>%
    separate(title, c("gene", "junk")) %>%
    mutate(gene = toupper(gene)) %>%
    select(-junk)

ecoli_features <- "training/Daley_gfp.fna.allstats.csv" %>%
    read.csv(as.is = TRUE, header = TRUE) %>%
    mutate(title = getname(title) %>% toupper)

fluman_asd <- fluman_asd %>%
    inner_join(ecoli_features, by = c("gene" = "title")) %>%
    inner_join(fluman_asd_features, by = c("gene" = "gene")) %>%
    mutate(rel_relareaSD = rel_diff(relareaSD.y, relareaSD.x),
           rel_totalSDsites = rel_diff(totalSDsites.y, totalSDsites.x))

p4_fluman_asd_prep <- fluman_asd %>%
    select(gene, rel_outcome, rel_ml21, rel_totalSDsites) %>% # ,
    gather(metric, value, -gene) %>%
    mutate(gene = factor(gene,
                         levels = c("YGDD", "BRNQ", "YBJJ"),
                         labels = c("YgdD", "BrnQ", "YbjJ")),
           metric = factor(metric,
                           levels = c("rel_totalSDsites", "rel_ml21", "rel_outcome"), #"rel_totalSDsites"), 
                           labels = c("SD-like Sites", "SVMrank Score", "Folded Protein"))) # "SD-like Elements", 

p4_fluman_asd <- ggplot(p4_fluman_asd_prep, aes(x = gene, y = value, fill = metric)) +
    geom_bar(stat = "identity", position = "dodge") + 
    ylab("Relative Difference") +
    scale_fill_manual(values = brewer.pal(3, "Dark2")) +
    scale_y_continuous(labels = c(-1, -.5, 0, .25)) + 
    theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key = element_blank(),
        legend.key.size = unit(0.25, "cm"),
        legend.position = c(.60, .20),
        aspect.ratio = 1)
```

## deal with feature distributions
```{r}
forward_features <- forward_preds %>%
    select(-ml21, -group) %>%
    mutate(set = "forward")

ecoli_features <- read.csv("training/Daley_gfp.fna.allstats.csv", as.is = TRUE, header = TRUE) %>%
    mutate(species = "Ec-Daley Training",
           set = "ecoli")
ecoli_features <- ecoli_features[, colnames(forward_features)]

nycomps_features <- rbind(read.csv("large-scale/nycomps_only_pos.fna.allstats.csv", as.is = TRUE, header = TRUE),
                          read.csv("large-scale/nycomps_mixed.fna.allstats.csv", as.is = TRUE, header = TRUE),
                          read.csv("large-scale/nycomps_only_neg.fna.allstats.csv", as.is = TRUE, header = TRUE)) %>%
    mutate(species = "nycomps all",
           set = "nycomps")
nycomps_features <- nycomps_features[, colnames(forward_features)]

# Don't need to recapitulate scaling and centering

all_features <- rbind(nycomps_features, ecoli_features, forward_features) %>%
    select(-title) %>%
    gather(featuretype, featurevalue, -set, -species) %>%
    mutate(thermophile = species %in% c("Aquifex aeolicus", "Spirochaeta thermophila", "Thermus thermophilus"))

estimate_overlap <- function (A, B, method = 'locfit', n.grid = 2^13) {
    # if the length of the sets are non-zero, the calculate a KDE
    if(length(A) == 0 | length(B) == 0) {
        return(NA)
    }
    
    # Calculate Extent
    range_all <- c(min(A, B), max(A, B))
    # Set up mesh
    mesh <- seq(range_all[[1]], range_all[[2]], length = n.grid)
    
    if(method == 'gss') {
        # fit each density independently, set extent with `domain`
        fitA <- gss::ssden(~A, domain=data.frame(A=range_all))
        fitB <- gss::ssden(~B, domain=data.frame(B=range_all))
        
        # calculate density at each grid point
        densityA <- gss::dssden(fitA, mesh)
        densityB <- gss::dssden(fitB, mesh)
    
        # normalize density to 1
        densityA <- densityA / sum(densityA)
        densityB <- densityB / sum(densityB)

    } else if(method == 'locfit') {
        # calculate density at each grid point
        densityA <- locfit::density.lf(A, ev = mesh, maxit = 10000)$y
        densityB <- locfit::density.lf(B, ev = mesh, maxit = 10000)$y
        
        # normalize density to 1
        densityA <- densityA / sum(densityA)
        densityB <- densityB / sum(densityB)
    } else if(method == 'ed') {
        densityA <- ed(A, xgrid = n.grid, bounds = c(min(A, B), max(A, B)))
        densityB <- ed(B, xgrid = n.grid, bounds = c(min(A, B), max(A, B)))
    } else {
        stop("Unrecognized method")
    }

    # calculate overall overlap
    sum(pmin(densityA, densityB))
}

getoverlap_thermo <- function(name) {
    set1 <- filter(all_features, thermophile, featuretype == name, !is.na(featurevalue))$featurevalue
    set2 <- filter(all_features, set == "ecoli", featuretype == name, !is.na(featurevalue))$featurevalue
    estimate_overlap(set1, set2) %>% as.numeric
}

getoverlap_set_set <- function(name, settype1, settype2 = "ecoli") {
    set1 <- filter(all_features, set == settype1, featuretype == name, !is.na(featurevalue))$featurevalue
    set2 <- filter(all_features, set == settype2, featuretype == name, !is.na(featurevalue))$featurevalue
    estimate_overlap(set1, set2) %>% as.numeric
}

getoverlap_species_set <- function(name, speciesname, settype = "ecoli") {
    set1 <- filter(all_features, species == speciesname, featuretype == name, !is.na(featurevalue))$featurevalue
    set2 <- filter(all_features, set == settype, featuretype == name, !is.na(featurevalue))$featurevalue
    estimate_overlap(set1, set2) %>% as.numeric
}

getoverlap_species_species <- function(name, speciesname1, speciesname2) {
    set1 <- filter(all_features, species == speciesname1, featuretype == name, !is.na(featurevalue))$featurevalue
    set2 <- filter(all_features, species == speciesname2, featuretype == name, !is.na(featurevalue))$featurevalue
    estimate_overlap(set1, set2) %>% as.numeric
}

name <- 'freqens'
getoverlap_thermo('freqens')

# for(i in unique(all_features$featuretype)) {
#     print(c(i, getoverlap_thermo(i)))
# }

# temp <- filter(all_features, featuretype == 'tAI')
# temp2 <- npudens(~ temp)
# temp3 <- head(temp$featurevalue, n=100)

# overlap$ecoli <- mclapply(unique(all_features$featuretype), getoverlap_species, speciesname = "Escherichia coli", 
#                   mc.cores = 12, mc.preschedule = FALSE) %>% simplify2array

overlap <- data.frame(
    feature = unique(all_features$featuretype),
    thermo = mclapply(unique(all_features$featuretype), getoverlap_thermo, 
                      mc.cores = 20, mc.preschedule = FALSE) %>% simplify2array %>% as.numeric,
    nycomps = mclapply(unique(all_features$featuretype), getoverlap_set_set, settype1 = "nycomps", settype2 = "ecoli",
                       mc.cores = 20, mc.preschedule = FALSE) %>% simplify2array %>% as.numeric,
    forward = mclapply(unique(all_features$featuretype), getoverlap_set_set, settype1 = "forward", settype2 = "ecoli",
                       mc.cores = 20, mc.preschedule = FALSE) %>% simplify2array %>% as.numeric,
    pf = mclapply(unique(all_features$featuretype), getoverlap_species_set, speciesname = "Plasmodium falciparum", settype = "ecoli",
                  mc.cores = 20, mc.preschedule = FALSE) %>% simplify2array %>% as.numeric,
    ecoli = mclapply(unique(all_features$featuretype), getoverlap_species_set, speciesname = "Escherichia coli", settype = "ecoli",
                  mc.cores = 20, mc.preschedule = FALSE) %>% simplify2array %>% as.numeric,
    human_training = mclapply(unique(all_features$featuretype), getoverlap_species_set, speciesname = "Homo sapiens", settype = "ecoli",
                  mc.cores = 20, mc.preschedule = FALSE) %>% simplify2array %>% as.numeric,
    human_ec = mclapply(unique(all_features$featuretype), getoverlap_species_species, speciesname1 = "Homo sapiens", speciesname2 = "Escherichia coli",
                  mc.cores = 20, mc.preschedule = FALSE) %>% simplify2array %>% as.numeric,
    human_sc = mclapply(unique(all_features$featuretype), getoverlap_species_species, speciesname1 = "Homo sapiens", speciesname2 = "Saccharomyces cerevisiae",
                  mc.cores = 20, mc.preschedule = FALSE) %>% simplify2array %>% as.numeric,
    human_pp = mclapply(unique(all_features$featuretype), getoverlap_species_species, speciesname1 = "Homo sapiens", speciesname2 = "Pichia pastoris",
                  mc.cores = 20, mc.preschedule = FALSE) %>% simplify2array %>% as.numeric,
    stringsAsFactors = FALSE)

# Mean absolute deviations:
mad_nycomps_forward <- mean(abs(overlap$nycomps - overlap$forward), na.rm = TRUE) * 100
mad_pf_nycomps <- mean(abs(overlap$pf - overlap$nycomps), na.rm = TRUE) * 100
mad_thermo_nycomps <- mean(abs(overlap$thermo - overlap$nycomps), na.rm = TRUE) * 100

p4_ecoli_dist <- ggplot(overlap) + 
    geom_vline(xintercept = 0.75, size = 0.7, linetype = "dotted", color = "#00b1b9") +
    geom_point(aes(x=ecoli,y=1), shape=17, alpha = 0.5, size = 0.4) +
    scale_x_continuous(expand = c(0,0), limits=c(0,1), labels = function(x) x*100) +
    xlab("E. coli vs E. coli (Training Set)") +
    theme(axis.title.y = element_blank(),
              axis.text.y = element_blank(),
              axis.line.y = element_blank(),
              axis.ticks.y = element_blank())

selected_features <- data.frame(
    feature = c("seqLen", "GC", "tAI", "CPS", "relareaSD", "zeroto38avgRNAss",
              "pI", "membrCont", "numPosNormCyt", "delGallTMs", "len1_2loop", "avgRONN"),
    x = c(2700, 0.7, 0.36, 100, 0.29, 0.8, 12, 450, 0.9, 4.5, 400, 0.7),
    y = c( 125, 110,   75, 125,  150,  60, 75, 275, 125,  90, 375,  75), stringsAsFactors = FALSE)

selected_features <- gdata:::interleave(selected_features[1:6,], selected_features[7:12,]) %>% 
    inner_join(overlap) %>%
    mutate(featuretype = factor(feature, levels = selected_features$feature),
           plotlabel = paste0(specify_decimal(thermo, k=2), "\n", specify_decimal(pf, k=2)))

p2ext_feature_dist_prep <- filter(all_features, set == "ecoli" | thermophile | species == "Plasmodium falciparum", 
                                  featuretype %in% selected_features$feature) %>%
    mutate(group = "E. coli (Training Data)", 
           featuretype = factor(as.character(featuretype), levels = selected_features$feature)) %>%
    # cull some of the features so plots are more meaningful
    filter(featuretype != 'len1_2loop' | featuretype == 'len1_2loop' & featurevalue < 500) %>%
    filter(featuretype != 'seqLen' | featuretype == 'seqLen' & featurevalue < 3000)

p2ext_feature_dist_prep[p2ext_feature_dist_prep$thermophile,]$group <- "Thermophiles"
p2ext_feature_dist_prep[p2ext_feature_dist_prep$species == "Plasmodium falciparum",]$group <- "P. falciparum"
p2ext_feature_dist_prep$group <- factor(p2ext_feature_dist_prep$group, 
                                        levels = c("E. coli (Training Data)", "Thermophiles", "P. falciparum"))

# thisgrp <- "Thermophiles"
# thistype <- "tAI"

p2ext_feature_dist_prep <- foreach(thisgrp=unique(p2ext_feature_dist_prep$group), .combine = rbind) %dopar% {
    foreach(thistype=unique(p2ext_feature_dist_prep$featuretype), .combine = rbind) %do% {
            thisdata <- filter(p2ext_feature_dist_prep, featuretype == thistype, group == thisgrp)
            thisrange <- range(thisdata$featurevalue)
            nbins <- 50
            grid <- seq(thisrange[[1]], thisrange[[2]], by = (thisrange[[2]]-thisrange[[1]])/nbins)
            freqs <- cut(thisdata$featurevalue, breaks = grid, ordered_result = TRUE) %>% table %>% data.frame
            data.frame(
                group = thisgrp,
                featuretype = thistype,
                x_left = grid[1:length(grid)-1],
                count = freqs$Freq,
                density = freqs$Freq / sum(freqs$Freq),
                labels = freqs$.)
    }
}

p2ext_feature_dist_prep <- p2ext_feature_dist_prep %>%
    mutate(featuretype = factor(featuretype,
                                levels = c("seqLen", "pI", "GC",
                                           "membrCont", "tAI", "numPosNormCyt",
                                           "CPS", "delGallTMs", "relareaSD",
                                           "len1_2loop", "zeroto38avgRNAss", "avgRONN"),
                                labels = c("Sequence Length", "Isoelectric Point", "GC Content", "Membrane Residue Count",
                                           "tRNA Adapation Index", "Positive Cytoplasmic Residues\n(normalized)", "Codon Pair Score",
                                           "ΔG of Insertion (Hessa, et al.)", "Shine-Dalgarno Sites (normalized)", "Length of TM1-TM2 Loop",
                                           "RNA Secondary Structure\n(1st 40 codons)", "Mean Disorder Prediction (RONN)")))

selected_features <- selected_features %>% 
    mutate(featuretype = factor(featuretype,
                                levels = c("seqLen", "pI", "GC",
                                           "membrCont", "tAI", "numPosNormCyt",
                                           "CPS", "delGallTMs", "relareaSD",
                                           "len1_2loop", "zeroto38avgRNAss", "avgRONN"),
                                labels = c("Sequence Length", "Isoelectric Point", "GC Content", "Membrane Residue Count",
                                           "tRNA Adapation Index", "Positive Cytoplasmic Residues\n(normalized)", "Codon Pair Score",
                                           "ΔG of Insertion (Hessa, et al.)", "Shine-Dalgarno Sites (normalized)", "Length of TM1-TM2 Loop",
                                           "RNA Secondary Structure\n(1st 40 codons)", "Mean Disorder Prediction (RONN)")))

pext2_feature_dist <- ggplot(p2ext_feature_dist_prep) +
    geom_step_hist(aes(x = featurevalue, color = group), alpha=0.7, position = "identity", bins = 50) +
    geom_text(aes(x = x, y = y, label = specify_decimal(thermo, k=2)), size = 5*5/14, color = brewer.pal(3, "Dark2")[[2]],
              data = selected_features) + 
    geom_text(aes(x = x, y = y*.9, label = specify_decimal(pf, k=2)), size = 5*5/14, color = brewer.pal(3, "Dark2")[[3]],
              data = selected_features) + 
    scale_fill_manual(values = brewer.pal(3, "Dark2")) + 
    scale_color_manual(values = brewer.pal(3, "Dark2")) + 
    scale_y_continuous(expand = c(0, 0)) + 
    scale_x_continuous(expand = c(0, 0)) +
    xlab("Value") + ylab("Density") + 
    facet_wrap(~featuretype, ncol = 2, scales = "free") +
    theme(axis.title = element_blank(),
          strip.background = element_blank(),
          legend.position="top",
          legend.key.size = unit(0.25, "cm"),
          legend.background = element_blank(),
          legend.key = element_blank(),
          legend.title = element_blank(),
          legend.margin = unit(c(0, 0, -.0115, 0), "npc")
          )

pext2_thermo <- ggplot(overlap, aes(x=nycomps, y=thermo, label=feature)) +
    geom_hline(yintercept = 0.75, linetype = "dotted", size = 0.7, color = "#00b1b9") + 
    geom_vline(xintercept = 0.75, linetype = "dotted", size = 0.7, color = "#00b1b9") +
    geom_abline(color="grey", linetype="dashed") +
    geom_point(size=0.25) +
    annotate(geom = "text", label = specify_decimal(mad_thermo_nycomps, k = 1), x = 0.1, y = 0.9, size = 6*5/14) + 
    #geom_text(size = 5*5/14, nudge_y = .005) +
#    geom_segment(aes(x = thermo, xend = thermo, y = nycomps, yend = forward), color = "black", size = 0.25,
#                 arrow = arrow(angle = 30, length = unit(1.25, "pt"), type = "closed"), data = overlap) +
    xlab("NYCOMPS vs. Training") +
    ylab("Thermophiles vs. Training") +
    scale_x_continuous(limits = c(0,1), expand = c(0,0), labels = function(x) x*100) +
    scale_y_continuous(limits = c(0,1), expand = c(0,0), labels = function(x) x*100) +
    theme(legend.position="none", legend.title = element_blank(), aspect.ratio = 1)

pext2_pf <- ggplot(overlap, aes(x=nycomps, y=pf, label=feature)) +
    geom_hline(yintercept = 0.75, linetype = "dotted", size = 0.7, color = "#00b1b9") + 
    geom_vline(xintercept = 0.75, linetype = "dotted", size = 0.7, color = "#00b1b9") +
    geom_abline(color="grey", linetype="dashed") +
    geom_point(size=0.25) +
    annotate(geom = "text", label = specify_decimal(mad_pf_nycomps, k = 1), x = 0.1, y = 0.9, size = 6*5/14) + 
    #geom_text(size = 5*5/14, nudge_y = .005) +
    xlab("NYCOMPS vs. Training") +
    ylab("P. falciparum vs. Training") +
    scale_color_manual(values = brewer.pal(3, "Dark2")[2:1]) + 
  #  scale_linetype_manual(values = 1) +
    scale_x_continuous(limits = c(0,1), expand = c(0,0), labels = function(x) x*100) +
    scale_y_continuous(limits = c(0,1), expand = c(0,0), labels = function(x) x*100) +
    theme(legend.position="none", legend.title = element_blank(), aspect.ratio = 1)

pext2_nycomps_forward <- ggplot(overlap, aes(x=nycomps, y=forward, label=feature)) +
    geom_hline(yintercept = 0.75, linetype = "dotted", size = 0.7, color = "#00b1b9") + 
    geom_vline(xintercept = 0.75, linetype = "dotted", size = 0.7, color = "#00b1b9") +
    geom_abline(color="grey", linetype="dashed") +
 #   geom_segment(x=.75, xend=.75, y = 0, yend = .75) + 
    geom_point(size = 0.25) + 
    annotate(geom = "text", label = specify_decimal(mad_nycomps_forward, k = 1), x = 0.1, y = 0.9, size = 6*5/14) + 
#    geom_text(size = 5*5/14, nudge_y = .005) +
    xlab("NYCOMPS vs. Training") +
    ylab("Forward vs. Training") +
    scale_x_continuous(limits = c(0,1), expand = c(0,0), labels = function(x) x*100) +
    scale_y_continuous(limits = c(0,1), expand = c(0,0), labels = function(x) x*100) +
    theme(aspect.ratio = 1)

pext2_hs_ec_sc <- ggplot(overlap, aes(x=human_ec, y=human_sc, label=feature)) +
    geom_hline(yintercept = 0.75, linetype = "dotted", size = 0.7, color = "#00b1b9") + 
    geom_vline(xintercept = 0.75, linetype = "dotted", size = 0.7, color = "#00b1b9") +
    geom_abline(color="grey", linetype="dashed") +
 #   geom_segment(x=.75, xend=.75, y = 0, yend = .75) + 
    geom_point(size = 0.25) + 
#    geom_text(size = 5*5/14, nudge_y = .005) +
    xlab("Human vs. E. coli") +
    ylab("Human vs. S. c") +
    scale_x_continuous(limits = c(0,1), expand = c(0,0), labels = function(x) x*100) +
    scale_y_continuous(limits = c(0,1), expand = c(0,0), labels = function(x) x*100) +
    theme(aspect.ratio = 1)

pext2_hs_ec_pp <- ggplot(overlap, aes(x=human_ec, y=human_pp, label=feature)) +
    geom_hline(yintercept = 0.75, linetype = "dotted", size = 0.7, color = "#00b1b9") + 
    geom_vline(xintercept = 0.75, linetype = "dotted", size = 0.7, color = "#00b1b9") +
    geom_abline(color="grey", linetype="dashed") +
 #   geom_segment(x=.75, xend=.75, y = 0, yend = .75) + 
    geom_point(size = 0.25) + 
#    geom_text(size = 5*5/14, nudge_y = .005) +
    xlab("Human vs. E. coli") +
    ylab("Human vs. P. pastoris") +
    scale_x_continuous(limits = c(0,1), expand = c(0,0), labels = function(x) x*100) +
    scale_y_continuous(limits = c(0,1), expand = c(0,0), labels = function(x) x*100) +
    theme(aspect.ratio = 1)


p4_main <- ggdraw() + 
    draw_plot(plot_grid(p4_selected_violin, p4_nvector_ppv, nrow = 2, align = "hv", rel_heights = c(3, 1),
                        labels = letters[1:2], label_size = 8), 
              x = -.01, y = 0, width = .72, height = 1) +
    draw_plot(plot_grid(p4_ecoli_dist, pext2_nycomps_forward, pext2_thermo, pext2_pf, p4_fluman_asd,
                        labels = letters[3:7], label_size = 8, ncol = 1, align = "hv",
                        rel_heights = c(1, 3, 3, 3, 3)),
              x = .72, y = 0, width = .27, height = 1) 
#    draw_label("NYCOMPS vs. E. coli (Training) -> Forward vs. E. coli (Training)", angle = 90, size = 7,
#               x = .73, y = .33) +
#    draw_plot(p4_selected_violin, x = 0, y = .25, width = .72, height = .79) +
#    draw_plot_label(letters[1:3], c(0, 0, 0.33), c(1, .28, .28), size = 8)

#pext2_thermo

save_plot(filename = "plots/fig4_fluman_thermo.pdf", plot = p4_main,
          base_width = uconv(136, "mm", "in", "Length"),
          base_height = 5.5)

pext2 <- plot_grid(p4_all_violin, pext2_feature_dist, rel_widths = c(2, 1), labels = letters[1:2], label_size = 8, ncol = 2)

# RGB by default
save_plot(filename = "plots/extfig2_names.pdf", 
          plot = pext2, 
          base_width = uconv(183, "mm", "in", "Length"), base_height = 7,
          dpi = 300)

```

# plot scores of features along with correlations within
# ecoli training, nycomps, and forward genomes datasets
```{r feature_plot}
# load preprocessing script
# features we use for the ML
load("~/homedir/ml-ecoli-paper/predict.RData")
preprocess_vars <- function(test_data, xtrans) {
    prediction_features <- names(xtrans$mean)
    
    # features in the test dataset
    test_data_features <- colnames(test_data)
    
    # features requested for the ML but not in the input dataset
    # keep in mind setdiff gives the asymmetric difference
    # (http://stat.ethz.ch/R-manual/R-patched/library/base/html/sets.html)
    missing_features <- setdiff(prediction_features, colnames(test_data))
    
    for(x in missing_features) {
        test_data[[x]] <- NA
    }
    rm(x)
    
    # keep only the columns we want for the ML (get rid of the extras)
    test_data <- test_data[, prediction_features]
    
    # do the preprocessing (scaling and centering)
    transformed <- predict(xtrans, test_data)
    
    transformed[, !(colnames(transformed) %in% missing_features)]
}

svmweights <- read.table("daley_only_raw_exp_ml21.full.4.svmweights", header = FALSE)
colnames(svmweights) <- c("feature", "weight")
svmweights$rel <- svmweights$weight / mean(svmweights$weight)
svmweights <- arrange(svmweights, desc(weight))

forward_cor <- preprocess_vars(forward_features,
                               daley_allstats_exp_xtrans_params) %>% 
    cor(use = "pairwise.complete.obs")

cor.mtest <- function(mat, conf.level = 0.95){
  mat <- as.matrix(mat)
    n <- ncol(mat)
    p.mat <- lowCI.mat <- uppCI.mat <- matrix(NA, n, n)
    diag(p.mat) <- 0
    diag(lowCI.mat) <- diag(uppCI.mat) <- 1
    for(i in 1:(n-1)){
        for(j in (i+1):n){
            tmp <- cor.test(mat[,i], mat[,j], conf.level = conf.level)
            p.mat[i,j] <- p.mat[j,i] <- tmp$p.value
            lowCI.mat[i,j] <- lowCI.mat[j,i] <- tmp$conf.int[1]
            uppCI.mat[i,j] <- uppCI.mat[j,i] <- tmp$conf.int[2]
        }
    }
    return(list(p.mat, lowCI.mat, uppCI.mat))
}
res1 <- cor.mtest(preprocess_vars(forward_features,
                                  daley_allstats_exp_xtrans_params))
pdf("plots/forward_corr.pdf")
corrplot(forward_cor, col=viridis(300),
         method="square", plotCI="n", order ="hclust",
         hclust.method="complete", addrect=11,
         p.mat = res1[[1]], sig.level=0.05, 
         insig = "blank", tl.cex=.5, tl.col="black", tl.srt=45, 
         lowCI.mat = res1[[2]], uppCI.mat = res1[[3]])
dev.off()

feat_order <- hclust(as.dist(1 - forward_cor), method = "complete")
feat_order <- feat_order$labels[feat_order$order]

svmweights_plot <- svmweights %>%
    filter(feature %in% feat_order) %>%
    mutate(feature = factor(feature, label = feat_order))

p_weights <-
    ggplot(svmweights_plot) +
    geom_bar(aes(x=feature, y=weight), stat="identity") +
    ylab(expression("SVM"^"rank"*~"score")) + 
    coord_flip() +
    theme(axis.title.y = element_blank())
```

```{r pca}

source("https://bioconductor.org/biocLite.R")
library(pcaMethods)

# replace NaN with NA
forward_trans <- preprocess_vars(forward_features,
                               daley_allstats_exp_xtrans_params) %>%
    as.matrix(forward_trans)
forward_trans[is.nan(forward_trans)] <- NA
forward_trans <- data.frame(forward_trans)

forward_nipals <- pca(forward_trans, center=TRUE, scale = "uv", method="nlpca", nPcs=20, completeObs = FALSE)
```

# Calculate Pfam overlap between E. coli and NYCOMPS
```{r}
ecoli_pfam <- rbind(
    read.table("../ml-datasets/ecoli-daley-fluman/precalc/Daley_gfp.fna.faa.pfamA",
               comment.char = '#', as.is = TRUE),
    read.table("../ml-datasets/ecoli-daley-fluman/precalc/Daley_phoa.fna.faa.pfamA",
               comment.char = '#', as.is = TRUE))

colnames(ecoli_pfam) <- c('seq_id', 'alignment_start', 'alignment_end',
                          'envelope_start', 'envelope_end', 'hmm_acc',
                          'hmm_name', 'type', 'hmm_start', 'hmm_end',
                          'hmm_length', 'bit_score', 'E-value',
                          'significance', 'clan')

nycomps_pfam <- rbind(
    read.table("../ml-datasets/nycomps/nycomps_only_pos.fna.faa.pfamA",
               comment.char = '#', as.is = TRUE),
    read.table("../ml-datasets/nycomps/nycomps_only_neg.fna.faa.pfamA",
               comment.char = '#', as.is = TRUE),
    read.table("../ml-datasets/nycomps/nycomps_mixed.fna.faa.pfamA",
               comment.char = '#', as.is = TRUE))
nycomps_pfam$id <- getname(nycomps_pfam$seq_id, token = 3) %>% as.numeric
colnames(nycomps_pfam) <- colnames(ecoli_pfam)

nycomps_pfam$hmm_training <- nycomps_pfam$hmm_acc %in% ecoli_pfam$hmm_acc
nycomps_pfam$clan_training <- nycomps_pfam$clan %in% ecoli_pfam$clan

nycomps_vectors$hmm_training <- nycomps_vectors$id %in% nycomps_pfam[nycomps_pfam$hmm_training,]$id
nycomps_vectors$clan_training <- nycomps_vectors$id %in% nycomps_pfam[nycomps_pfam$clan_training,]$id

nycomps_auc_hmm_training <- foreach(thisplasmid = unique(nycomps_vectors$plasmid_name), .combine=rbind) %dopar% {
    thisdata <- filter(nycomps_vectors, plasmid_name == thisplasmid, hmm_training)
    thisdata <- transmute(thisdata, 
                          thisresponse = plasmid_outcome == 1,
                          ml21 = ml21)
    df1 <- data.frame(
        ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
        group = thisplasmid,
        type = "plasmid",
        cterm = "all",
        ci_labels = c("lower95", "auc", "upper95"),
        count = nrow(thisdata),
        pos_count = sum(thisdata$thisresponse)
        )
    
    df2 <- foreach(thiscterm = unique(nycomps_vectors$cterm), .combine=rbind) %dopar% {
        thisdata <- filter(nycomps_vectors, plasmid_name == thisplasmid, cterm == thiscterm)
        thisdata <- transmute(thisdata, 
                              thisresponse = plasmid_outcome == 1,
                              ml21 = ml21)
        data.frame(
            ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
            group = thisplasmid,
            type = "plasmid",
            cterm = thiscterm,
            ci_labels = c("lower95", "auc", "upper95"),
            count = nrow(thisdata),
            pos_count = sum(thisdata$thisresponse)
            )
    }
    
    rbind(df1, df2)
} %>% mutate(hmm = TRUE)
nycomps_auc_hmm_training <- foreach(thisplasmid = unique(nycomps_vectors$plasmid_name), .combine=rbind) %dopar% {
    thisdata <- filter(nycomps_vectors, plasmid_name == thisplasmid, !hmm_training)
    thisdata <- transmute(thisdata, 
                          thisresponse = plasmid_outcome == 1,
                          ml21 = ml21)
    df1 <- data.frame(
        ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
        group = thisplasmid,
        type = "plasmid",
        cterm = "all",
        ci_labels = c("lower95", "auc", "upper95"),
        count = nrow(thisdata),
        pos_count = sum(thisdata$thisresponse)
        )
    
    df2 <- foreach(thiscterm = unique(nycomps_vectors$cterm), .combine=rbind) %dopar% {
        thisdata <- filter(nycomps_vectors, plasmid_name == thisplasmid, cterm == thiscterm)
        thisdata <- transmute(thisdata, 
                              thisresponse = plasmid_outcome == 1,
                              ml21 = ml21)
        data.frame(
            ml21 = ci.auc(thisresponse ~ ml21, data = thisdata, direction = "<", method = "delong"),
            group = thisplasmid,
            type = "plasmid",
            cterm = thiscterm,
            ci_labels = c("lower95", "auc", "upper95"),
            count = nrow(thisdata),
            pos_count = sum(thisdata$thisresponse)
            )
    }
    
    rbind(df1, df2)
} %>% mutate(hmm = FALSE) %>% rbind(nycomps_auc_hmm_training)

nycomps_auc_hmm_training <- dcast(nycomps_auc_hmm_training, group + type + cterm + count + pos_count + hmm ~ ci_labels, value.var = "ml21")

ggplot(nycomps_auc_hmm_training %>% filter(cterm == "all")) +
    geom_errorbar(aes(x = group,
                   ymin = upper95,
                   ymax = lower95),
              width = 0.25) + 
    geom_point(aes(x = group, y = auc, color = hmm))
    # geom_segment(aes(x = parent_score,
    #                xend = mutant_score,
    #                y = parent_expression,
    #                yend = mutant_expression,
    #                color = parent), 
    #            arrow = arrow(length = unit(0.02, "npc"), type = "closed"))
```

```{r}
sessionInfo()
```
